<!DOCTYPE html>
<html>
<head>
<title>Crux Tab-delimited File Format</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-V7XKGGFPYX"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-V7XKGGFPYX');
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
	// Main Menu
	$( document ).ready(function() {
			var pull 		= $('.btn');
				menu 		= $('nav ul');
				menuHeight	= menu.height();

			$(pull).on('click', function(e) {
				e.preventDefault();
				menu.slideToggle();
			});

			$(window).resize(function(){
	    		var w = $(window).width();
	    		if(w > 320 && menu.is(':hidden')) {
	    			menu.removeAttr('style');
	    		} 
			});
	});
</script>
</head>
<body>
	<div class="page-wrap">
	 	<nav>
			<div class="btn">
		    </div>
		    <img src="../images/crux-logo.png" id="logo"></a>
			<ul id="navitems">
			    <li><a href="../index.html">Home</a></li>
			    <li><a href="../download.html">Download</a></li>
			    <li><a href="../fileformats.html">File Formats</a></li>
		        <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
		    </ul>
		</nav>
		<div id="content">
		<!-- START CONTENT -->
		<h2>Tab-delimited file format</h2>
		<p>
		Various programs in Crux report their output and read inputs in tab-delimited text format. Each such file consists of a single line containing the tab-separated names of fields, followed by one or more lines giving the corresponding field values.
		</p>
		<table border="1" id="tabformat">
		<!-- table headers-->
		<tr>
		<th rowspan="2">Name</th>
		<th rowspan="2">Description</th>
 		<th rowspan="2">comet</th>
		<th rowspan="2">tide</th>
		<th rowspan="2">cascade-search</th>
    <th rowspan="2">kojak</th>
		<th colspan="3">percolator</th>
		<th rowspan="2">diameter search</th>
		<th rowspan="2">spectral-counts</th>
		</tr>

		<tr>
		<!--percolator-->
		<th>PSM</th>
		<th>peptides</th>
		<th>proteins</th>
		</tr>

		<!--"✓"= "&#10003;" or "&#10004"-->
		<!--"✕"= "&#10005" or "&#10006"-->
		<!--"–"=&#8211-->

		<tr>
		<td>file</td>
		<td>The name of the file containing the scan.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>scan</td>
		<td>The identifying number of each scan.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>charge</td>
		<td>The charge state for this spectrum in this PSM.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>spectrum precursor m/z</td>
		<td>The observed m/z of the spectrum precursor ion.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>spectrum neutral mass</td>
		<td>The computed mass of the spectrum precursor at the given charge
		state. This is equal to the precursor m/z minus the mass of a proton
		(1.00727646677 Da), all multiplied by the charge.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>peptide mass</td>
		<td>The mass of the peptide sequence, computed as the sum of the
		amino acid masses plus the mass of water (18.010564684 Da or
		18.0153 Da, depending on whether we are using monoisotopic or
		average mass).</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>delta_cn</td>
		<td>The normalized difference in XCorr for this PSM
		relative to the next ranked PSM for the same spectrum
		and charge. The denominator for normalization is the
		maximum of the current XCorr and 1.0.
		If <code>exact-p-value=T</code>, then the difference
		is computed between -log(p-value) rather than XCorr,
		and no normalization is applied.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>delta_lcn</td>
		<td>Similar to delta_cn, except that the difference
		is computed with respect to the lowest reported XCorr
		score for a given spectrum and charge state.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>sp score</td>
		<td>The SEQUEST-type preliminary score.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>sp rank</td>
		<td>The rank of this PSM when sorted by Sp score.  Note that,
		in <code>tide-search</code>, the Sp score is only computed for PSMs
		that are reported to the user.  Hence, the rank of the Sp score
		will be in the range from 1 to <i>n</i>, where the value
		of <i>n</i> is determined by the <code>--top-match</code> parameter.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>xcorr score</td>
		<td>The SEQUEST-type cross correlation score.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>exact p-value</td>
		<td>The p-value computed as described in
		<a href="http://www.ncbi.nlm.nih.gov/pubmed/24895379">"Computing
		Exact p-values for a Cross-correlation Shotgun
		Proteomics Score Function."</a>
		</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>refactored xcorr</td>
		<td>A discretized version of XCorr, used to compute
		the p-value. </td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>res-ev p-value</td>
		<td>The high-resolution p-value computed as described in
		<a href="https://pubs.acs.org/doi/abs/10.1021/acs.jproteome.8b00206">
		"Combining High-Resolution and Exact Calibration to
		Boost Statistical Power: A Well-Calibrated Score
		Function for High-Resolution MS2 Data."</a></td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>res-ev score</td>
		<td>The high-resolution score computed as described in
		<a href="https://pubs.acs.org/doi/abs/10.1021/acs.jproteome.8b00206">
		"Combining High-Resolution and Exact Calibration to
		Boost Statistical Power: A Well-Calibrated Score
		Function for High-Resolution MS2 Data."</a>
		</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>combined p-value</td>
		<td>The high-resolution p-value computed as described in
		<a href="https://pubs.acs.org/doi/abs/10.1021/acs.jproteome.8b00206">
		"Combining High-Resolution and Exact Calibration to
		Boost Statistical Power: A Well-Calibrated Score
		Function for High-Resolution MS2 Data."</a>
		</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

                <tr>
		<td>tailor score</td>
		<td>A calibrated version of the XCorr score,
		normalized by dividing by the 99th percentile of the
		XCorr scores for each spectrum, as described in
		<a href="https://pubmed.ncbi.nlm.nih.gov/32175744/">
		"Tailor: A Nonparametric and Rapid Score Calibration
		Method for Database Search-Based Peptide
		Identification in Shotgun Proteomics."</a></td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>xcorr rank</td>
		<td>The rank of this PSM when sorted by xcorr.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>res-ev rank</td>
		<td>The rank of this PSM when sorted by res-ev score or by res-ev p-value.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>combined p-value rank</td>
		<td>The rank of this PSM when sorted by combined p-value.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>b/y ions matched</td>
		<td>The number of b-ions and y-ions in the peptide that have a
		corresponding peak in the spectrum.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>b/y ions total</td>
		<td>The total number of b- and y-ions predicted for this peptide.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>total matches/spectrum</td>
		<td>The number of candidate peptides in the database found for this
		spectrum, including duplicates.  Note that this is always the number
		of target candidate peptides, even if the PSM involves a decoy peptide.</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>distinct matches/spectrum</td>
		<td>The number of unique candidate peptides in the database found for this
		spectrum.  Note that this is always the number of target candidate
		peptides, even if the PSM involves a decoy peptide.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>sequence</td>
		<td>The peptide sequence.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>modifications</td>
		<td>Variable modifications applied to the sequence.
                  This is comma-separated list of modifications in the
                  format "position_code_massdiff",
                  e.g. "3_V_15.9949". The "position" field is the
                  position of the peptide residue, where position 1 is
                  the first residue. A position of "0" denotes the
                  previous flanking amino acid and a position of 1
                  greater than the peptide length denotes the
                  following flanking amino acid. The "code" field can
                  be "S" for a static modification or "V" for a
                  variable modification. The "modifications" string
                  can be appended with "_N" to denote an N-term
                  protein modification, e.g. "1_S_-17.0265_N"; "_n" to
                  denote an N-term peptide modification,
                  e.g. "1_A_42.0146_n"; "_C" to denote a C-term
                  protein modification, e.g. "9_R_356.1882_C"; or "_c"
                  to denote a C-term peptide modification,
                  e.g. "12_K_42.0106_c".</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>cleavage type</td>
		<td>The cleavage rules for generating this peptide based on the
		user-specified enzyme specificity.</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>unmodified sequence</td>
		<td>The peptide sequence stripped all the modification related information.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>protein id</td>
		<td>A comma-separated list of proteins in which this peptide
		appears. Optionally, the protein name may be followed by a number
		in parentheses giving the start position of the peptide in the
		protein.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>flanking aa</td>
		<td>The amino acids that precede and follow this
		peptide in the parent protein ID.  If the peptide
		occurs in more than one protein, then this column
		will contain a comma-separated list of pairs of amino
		acids.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>original target sequence</td>
		<td>The unmodified target sequence. For a target PSM,
		the value in this column will be identical to the
		value in the "sequence" column.  For a decoy PSM,
		this column will contain the corresponding target
		sequence.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>PSMId</td>
		<td>Identifier of this peptide-spectrum match.  If
		the PIN file was created by Crux, then the ID will be
		of the form target_0_8000_2_1, where the components
		are "target" or "decoy," the file index, scan number,
		charge, and PSM rank.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>score</td>
		<td>The discriminant score assigned by percolator.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>filename</td>
		<td>Name of the file in which the scan was found.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>q-value</td>
		<td>The q-value assigned by percolator.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>posterior_error_prob</td>
		<td>The posterior error probability assigned by percolator.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>peptide</td>
		<td>The peptide sequence.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>proteinIds</td>
		<td>A comma-separated list of proteins in which this peptide
		appears.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>proteinId</td>
		<td>Identifier for this protein or protein group</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>proteinGroupId</td>
		<td>Identifier associated with this protein group</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>peptidesIds</td>
		<td>Peptides belonging to each protein in the group</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>precursor intensity logrank M0</td>
		<td>The log-rank of precursor intensity (for
		monoisotope) among all observed peaks in MS1 scan</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>precursor intensity logrank M1</td>
		<td>The log-rank of precursor intensity (for M + 1
		isotope) among all observed peaks in MS1 scan</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>precursor intensity logrank M2</td>
		<td>The log-rank of precursor intensity (for M +2
		isotope) among all observed peaks in MS1 scan</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>rt-diff</td>
		<td>The difference between observed and predicted
		retention time</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>dynamic fragment p-value</td>
		<td>Fragment matching p-value, as described in
		<a href="https://pubs.acs.org/doi/10.1021/pr500202e">
		"MS Amanda, a Universal Identification Algorithm
		Optimized for High Accuracy Tandem Mass Spectra"</a>.
                When calculating the score, the m/z range
		(dependent on the observed peaks) is divided into 10
		equal-length segments and the 10 most intense peaks
		within each bin are preserved.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>static fragment p-value</td>
		<td>Similar to the static fragment p-value, except
		that the m/z range is set independently of the observed
		peaks.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>precursor coelution</td>
		<td>The normalized dot product between the elution
		profiles of precursor monoisotope and M+1 and M+2
		isotopes</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>fragment coelution</td>
		<td>The normalized dot product between the elution
		profiles of fragment ions</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>precursor fragment coelution</td>
		<td>The normalized dot product between the elution
		profiles of precursor isotopes and fragment ions</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>ensemble score</td>
		<td>The aggregated score calculated as the weighted
		sum of used features</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>target/decoy</td>
		<td>Is this peptide a target or a decoy? Value is
		"target" or "decoy."</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>RAW</td>
		<td>The raw (unnormalized) count of spectra per peptide</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>SIN</td>
		<td>A protein quantification score</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>NSAF</td>
		<td>The normalized spectral abundace factor</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>dNSAF</td>
		<td>The distributed normalized spectral abundace factor</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>EMPAI</td>
		<td>The exponentially modified protein abundance index</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td> parsimony rank</td>
		<td>The protein rank based on its spectral-counts score.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>Scan Number</td>
		<td>The identifying number of each scan.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Ret Time</td>
		<td>The retention time in minutes.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Obs Mass</td>
		<td>The observed mass of the precursor ion peak.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>PSM Mass</td>
		<td>The theoretically computed precursor mass of the PSM.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>PPM Error</td>
		<td>The difference (in parts per million) between the PSM mass and the Obs Mass.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Score</td>
		<td>The Kojak cross-correlation score.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>dScore</td>
		<td>The difference between the reported PSM score and the next best PSM score.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>E-value</td>
		<td>The overall E-value of the PSM.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Peptide #1 Score</td>
		<td>The Kojak xcorr score of the higher scoring peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Peptide #1 E-value</td>
		<td>The E-value of the higher scoring peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Peptide #1</td>
		<td>The peptide sequence of the higher scoring peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Linked AA #1</td>
		<td>The site of the cross-link in Peptide #1</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Protein #1</td>
		<td>The protein(s) that generated Peptide #1.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Protein #1 Site</td>
		<td>The site(s) of the cross-link in Protein #1</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Peptide #2 Score</td>
		<td>The Kojak xcorr score of the lower scoring peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Peptide #2 E-value</td>
		<td>The E-value of the lower scoring peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Peptide #2</td>
		<td>The peptide sequence of the lower scoring peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Linked AA #2</td>
		<td>The site of the cross-link in Peptide #2</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Protein #2</td>
		<td>The protein(s) that generated Peptide #2.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>Protein #2 Site</td>
		<td>The site(s) of the cross-link in Protein #2</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>Linker Mass</td>
		<td>The mass contribution of the cross-linker for cross-linked
		PSMs or zero otherwise.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>
		</table> 
<!-- END CONTENT -->
		</div>
	</div>

<footer class="site-footer">
	<div id="centerfooter">
		<div class="footerimportantlinks">
			<img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
			<ul>
				<li><a href="../faq.html">Crux FAQ</a></li>
				<li><a href="../glossary.html">Glossary of terminology</a></li>
				<li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
				<li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
				<li><a href="../release-notes.html">Release Notes</a></li>
				<li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
				<li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
				<li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
			</ul>
		</div>
		<div class="footerimportantlinks tutoriallinks">
			<img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
			<ul>
				<li><a href="../tutorials/install.html">Installation</a></li>
				<li><a href="../tutorials/gettingstarted.html">Getting started with Crux</a></li>
				<li><a href="../tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
				<li><a href="../tutorials/customizedsearch.html">Customization and search options</a></li>
				<li><a href="../tutorials/spectralcounts.html">Using spectral-counts</a></li>
			</ul>
		</div>
		<div id="footertext">
			<p>
				The original version of Crux was written by Chris Park and Aaron Klammer
				under the supervision
				of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
				MacCoss</a>
				and <a href="http://noble.gs.washington.edu/~noble">Prof. William
				Stafford Noble</a> in the Department of Genome Sciences at the
				University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
				<br />The complete list of contributors
				can be found <a href="../contributors.html">here</a>.
				<br />
				<br />
				Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
			</p>
		</div>
	</div>
</footer>
	
</body>

</html>
