<!DOCTYPE html>
<html>
<head>
<title>Crux Tab-delimited File Format</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
	// Main Menu
	$( document ).ready(function() {
			var pull 		= $('.btn');
				menu 		= $('nav ul');
				menuHeight	= menu.height();

			$(pull).on('click', function(e) {
				e.preventDefault();
				menu.slideToggle();
			});

			$(window).resize(function(){
	    		var w = $(window).width();
	    		if(w > 320 && menu.is(':hidden')) {
	    			menu.removeAttr('style');
	    		} 
			});
	});
</script>
</head>
<body>
	<div class="page-wrap">
	 	<nav>
			<div class="btn">
		    </div>
		    <img src="../images/crux-logo.png" id="logo"></a>
			<ul id="navitems">
			    <li><a href="../index.html">Home</a></li>
			    <li><a href="../download.html">Download</a></li>
			    <li><a href="../fileformats.html">File Formats</a></li>
		        <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
		    </ul>
		</nav>
		<div id="content">
		<!-- START CONTENT -->
		<h2>Tab-delimited file format</h2>
		<p>
		Various programs in Crux report their output and read inputs in tab-delimited text format. Each such file consists of a single line containing the tab-separated names of fields, followed by one or more lines giving the corresponding field values.
		</p>
		<table border="1" id="tabformat">
		<!-- table headers-->
		<tr>
		<th rowspan="2">Name</th>
		<th rowspan="2">Description</th>
		<th rowspan="2">comet</th>
		<th rowspan="2">tide</th>
		<th rowspan="2">cascade-search</th>
		<th rowspan="2">search-for-xlinks</th>
		<th rowspan="2">search-for-xlinks (new)</th>
		<th colspan="4">barista</th>
		<th rowspan="2">q-ranker</th>
		<th colspan="3">percolator</th>
		<th rowspan="2">spectral-counts</th>
		</tr>

		<tr>
		<!--barista-->
		<th>PSM</th>
		<th>peptides</th>
		<th>proteins</th>
		<th>sub-proteins</th>
		<!--percolator-->
		<th>PSM</th>
		<th>peptides</th>
		<th>proteins</th>
		</tr>

		<!--"✓"= "&#10003;" or "&#10004"-->
		<!--"✕"= "&#10005" or "&#10006"-->
		<!--"–"=&#8211-->

		<tr bgcolor="#DEDEDE">
		<td>file</td>
		<td>The name of the file containing the scan.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>file</td>
		<td>The integer index of the file containing the scan.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>scan</td>
		<td>The identifying number of each scan.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>charge</td>
		<td>The charge state for this spectrum in this PSM.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>spectrum precursor m/z</td>
		<td>The observed m/z of the spectrum precursor ion.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>spectrum neutral mass</td>
		<td>The computed mass of the spectrum precursor at the given charge   state. This is equal to the precursor m/z minus the mass of a proton (1.00727646677 Da), all multiplied by the charge.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>peptide mass</td>
		<td>The mass of the peptide sequence, computed as the sum of the the amino acid masses plus the mass of water (18.010564684 Da or 18.0153 Da, depending on whether we are using monoisotopic or average mass).</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>delta_cn</td>
		<td>The normalized difference in XCorr for this PSM
		relative to the next ranked PSM for the same spectrum
		and charge. The denominator for normalization is the
		minimum of the current XCorr and 1.0.
		If <code>exact-p-value=T</code>, then the difference
		is computed between -log(p-value) rather than XCorr,
		and no normalization is applied.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		  <td>delta_lcn</td>
		<td>Similar to delta_cn, except that the difference
		is computed with respect to the lowest reported XCorr
		score for a given spectrum and charge state.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>sp score</td>
		<td>The SEQUEST-type preliminary score.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>sp rank</td>
		<td>The rank of this PSM when sorted by Sp score.  Note that,
		  in <code>tide-search</code>, the Sp score is only computed for PSMs
		  that are reported to the user.  Hence, the rank of the Sp score
		  will be in the range from 1 to <i>n</i>, where the value
		  of <i>n</i> is determined by the <code>--top-match</code> parameter.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>xcorr score</td>
		<td>The SEQUEST-type cross correlation score.
		This column only appears when <code>exact-p-value=F</code>.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>exact p-value</td>
		<td>The p-value computed as described in
		  <a href="http://www.ncbi.nlm.nih.gov/pubmed/24895379">"Computing Exact p-values
		  for a Cross-correlation Shotgun Proteomics Score Function"</a>.
		  This column only appears when <code>exact-p-value=T</code>.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>refactored xcorr</td>
		<td>A discretized version of XCorr, used to compute the p-value.
		This column only appears when <code>exact-p-value=T</code>.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>xcorr rank</td>
		<td>The rank of this PSM when sorted by xcorr.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>p-value</td>
		<td>The p-value computed from the estimated Weibull parameters of the score distribution for this spectrum.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>percolator score</td>
		<td>The discriminant score assigned by percolator.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>percolator q-value</td>
		<td>The q-value assigned by percolator.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>percolator PEP</td>
		<td>The posterior error probability assigned by percolator.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>q-ranker score</td>
		<td>The discriminant score assigned by q-ranker.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>q-ranker q-value</td>
		<td>The q-value assigned by q-ranker.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>q-ranker PEP</td>
		<td>The posterior error probability estimated from the q-ranker score</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>barista score</td>
		<td>The discriminant score assigned by barista.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>barista q-value</td>
		<td>The q-value assigned by barista.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>barista PEP</td>
		<td>The PEP assigned by barista</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>b/y ions matched</td>
		<td>The number of b-ions and y-ions in the peptide that have a corresponding peak in the spectrum.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>b/y ions total</td>
		<td>The total number of b- and y-ions predicted for this peptide.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>total matches/spectrum</td>
		<td>The number of candidate peptides in the database found for this
		  spectrum, including duplicates.  Note that this is always the number of target candidate
		  peptides, even if the PSM involves a decoy peptide.</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>distinct matches/spectrum</td>
		<td>The number of unique candidate peptides in the database found for this
		  spectrum.  Note that this is always the number of target candidate
		  peptides, even if the PSM involves a decoy peptide.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>


		<tr bgcolor="#DEDEDE">
		<td>sequence</td>
		<td>The peptide sequence, not including any modifications.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>modified sequence</td>
		<td>The peptide sequence, with modifications included.</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		  <td>modifications</td>
                  <!-- This description needs to be expanded considerably.  -->
		<td>Variable modifications applied to the sequence.</td>
		<td></td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>cleavage type</td>
		<td>The cleavage rules for generating this peptide based on the user-specified enzyme specificity.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>protein id</td>
		<td>A comma-separated list of proteins in which this peptide
		  appears. Optionally, the protein name may be followed by a number in parentheses giving the start position of the peptide in the protein.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>flanking aa</td>
		<td>The amino acids that precede and follow this
		peptide in the parent protein ID.  If the peptide
		occurs in more than one protein, then this column
		will contain a comma-separated list of pairs of amino
		acids.</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>original target sequence</td>
		<td>The unmodified target sequence. For a target PSM,
		the value in this column will be identical to the
		value in the "sequence" column.  For a decoy PSM,
		this column will contain the corresponding target
		sequence.</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>RAW</td>
		<td>The raw (unnormalized) count of spectra per peptide</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>SIN</td>
		<td>A protein quantification score</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>NSAF</td>
		<td>The normalized spectral abundace factor</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003</td>
		<td>&#10003</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>dNSAF</td>
		<td>The distributed normalized spectral abundace factor</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>EMPAI</td>
		<td>The exponentially modified protein abundance index</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td> parsimony rank</td>
		<td>The protein rank based on its spectral-counts score.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		</tr>

		<tr>
		<td>psm_id</td>
		<td>PSM identifier.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>protein group number</td>
		<td>Protein group index</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>peptides-scan.charge</td>
		<td>Peptides belonging to each protein in the group, with scan and charge (for example, EAMPK-001285.2)</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>proteins</td>
		<td>Proteins in the protein group. Each protein can be followed by peptides in parenthesis, if they belong only to this particular protein in the group. Since peptides are considered indistinguishable if they have identical amino acid sequences or they differ only by I/L or T/S in the same position in the peptide, the peptides shared by the group may not all be identical.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>parent group numbers</td>
		<td>The numbers of the protein groups in the barista.target.proteins.txt that are supersets of this group in terms of peptides.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>peptide mass mono</td>
		<td>Monoisotopic peptide mass</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>peptide mass average</td>
		<td>Average peptide mass</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>mass error(ppm)</td>
		<td>The mass error (in ppm) between the observed precursor mass and the
		    cross-link peptide product</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>protein id(loc)</td>
		<td>List of protein IDs that give rise to the first and second peptide in the
		    cross-link peptide product. (loc) indicates the location of the
		    peptide in the protein.  The ";" separates the protein ids from the
		    first and second peptide.</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>


		<tr>
		<td>protein id(loc) 1</td>
		<td>List of protein IDs that give rise to the first peptide in the
		    cross-link peptide product. (loc) indicates the location of the
		    peptide in the protein</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>protein id(loc) 2</td>
		<td>List of protein IDs that give rise to the second peptide in the
		    cross-link peptide product. (loc) indicates the location of the
		    peptide in the protein</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>by total</td>
		<td>Total number of b-y ions in product</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>by observable (0-1200)</td>
		<td>Total number of b and y ions observable in the 0-1200 m/z range</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>by observable bin (0-1200)</td>
		<td>Total number of b and y ions observable in the 0-1200 m/z range
		  using 1 Da bins</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>by observable (0-max)</td>
		<td>Total number of b and y ions observable in the entire m/z range</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>by observable bin (0-max)</td>
		<td>Total number of b and y ions observable in the entire m/z range
		  using 1 Da bins</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>by observed bin</td>
		<td>Total number of b-y ions observed in spectrum</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>ion current total</td>
		<td>Total ion current for spectrum</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr bgcolor="#DEDEDE">
		<td>ion current observed</td>
		<td>Total ion current observed for matching b-y ions</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		<tr>
		<td>ions observable bin (0-1200)</td>
		<td>Total ions observable (b, y, flanks, and neutral losses) within 0-1200 m/z range</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&#10003;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		<td>&nbsp;</td>
		</tr>

		</table> 
<!-- END CONTENT -->
		</div>
	</div>

<footer class="site-footer">
	<div id="centerfooter">
		<div class="footerimportantlinks">
			<img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
			<ul>
				<li><a href="../faq.html">Crux <strong>FAQ</strong></a></li>
				<li><a href="../glossary.html">Glossary of terminology</a></li>
				<li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
				<li><a href="https://sourceforge.net/projects/cruxtoolkit/">SourceForge Issue's list</a></li>
				<li><a href="../release-notes.html">Release Notes</a></li>
				<li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
				<li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
				<li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
			</ul>
		</div>
		<div class="footerimportantlinks tutoriallinks">
			<img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
			<ul>
				<li><a href="../tutorials/install.html">Installation</a></li>
				<li><a href="../tutorials/gettingstarted.html">Getting started with Crux</a></li>
				<li><a href="../tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
				<li><a href="../tutorials/customizedsearch.html">Customization and search options</a></li>
				<li><a href="../tutorials/spectralcounts.html">Using spectral-counts</a></li>
			</ul>
		</div>
		<div id="footertext">
			<p>
				The original version of Crux was written by Chris Park and Aaron Klammer
				under the supervision
				of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
				MacCoss</a>
				and <a href="http://noble.gs.washington.edu/~noble">Prof. William
				Stafford Noble</a> in the Department of Genome Sciences at the
				University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
				<br />The complete list of contributors
				can be found <a href="../contributors.html">here</a>.
				<br />
				<br />
				Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> awards R01 GM096306 and P41 GM103533. 
			</p>
		</div>
	</div>
</footer>
	
</body>

</html>
