<!DOCTYPE html>
<html>
<head>
<title>Getting started with Crux</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
	// Main Menu
	$( document ).ready(function() {
			var pull 		= $('.btn');
				menu 		= $('nav ul');
				menuHeight	= menu.height();

			$(pull).on('click', function(e) {
				e.preventDefault();
				menu.slideToggle();
			});

			$(window).resize(function(){
	    		var w = $(window).width();
	    		if(w > 320 && menu.is(':hidden')) {
	    			menu.removeAttr('style');
	    		} 
			});
	});
</script>
</head>
<body>
	<div class="page-wrap">
	 	<nav>
			<div class="btn">
		    </div>
		    <img src="../images/crux-logo.png" id="logo"></a>
			<ul id="navitems">
			    <li><a href="../index.html">Home</a></li>
			    <li><a href="../download.html">Download</a></li>
			    <li><a href="../fileformats.html">File Formats</a></li>
		        <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
		    </ul>
		</nav>
		<div id="content">
		<!-- START CONTENT -->
		<h2>Getting started with Crux</h2>

		<p> This page will talk about setting up your environment and choosing
		your input files.  Be sure to have completed the software <a
		href="install.html">installation</a> first.
		</p>

		<a name="env"></a>
		<h3>Your environment</h3>
		<p>For this tutorial, we'll assume you are working in a Linux/Unix
		type shell (<a href="http://www.cygwin.com/">cygwin</a> is a good
		choice for Windows users) and already know some basic 
		commands for changing directories, listing files, and other simple
		tasks.  To successfully run the sample commands, you'll need to work from 
		a directory for which you have write permission. Anywhere
		in your home directory should work.  Create a new directory and
		navigate to it.</p> 
		<div class=code><code>
		$ mkdir crux-demo<br>
		$ cd crux-demo</code>
		</div>
		<p>
		We will refer to this directory, <code>'crux-demo'</code>, as the
		<em>working directory</em>.</p> 
		<p>Next, make sure that the computer knows where to look for the
		<span class=prog>crux</span> programs.  Try this command. </p>
		<div class=code><code> 
		$ which crux</code>
		</div>
		<p>
		If it returns a single line with a path ending in <span
		class=out>crux</span>, then you are set.  If not,
		review the <a href="install.html#path">installation instructions</a> on setting your
		<span class=code>$PATH</span> environment variable.

		</p>

		<a name=ms2></a>
		<h3>Input file: mass spectra</h3>
		<p>Included in the <span class=prog>crux</span> distribution in the
		<span class=file>doc/user/data</span> directory are some 
		sample files containing mass spectra.  We will use the <a
		href="../example-files/demo.ms2">demo.ms2</a> for this tutorial.  Locate the
		file and copy it to the current working directory.
		</p>
		<p>The beginning of the file looks like this.
		</p>
		<pre>
H       CreationDate    2/14/2007 6:19:18 PM
H       Extractor       MakeMS2
H       ExtractorVersion        1.0
H       Comments        MakeMS2 written by Michael J. MacCoss, 2004
H       ExtractorOptions        MS2/MS1
S       10      10      636.34
Z       2       1271.67
187.4 12.5
193.1 19.5
194.3 13.7
		</pre>
		<div class=note><em>NOTE: There are two kinds of optional lines which
		may be included for each spectrum.  Lines beginning with </em><span
		class=file>I</span><em> are contain information independent of the
		charge state. Following a </em> <span class=file>Z</span><em> line may
		be one beginning with </em><span class=file>D</span><em> containing
		information specific to the above charge state. </em>
		</div>

		<p>
		The first lines beginning with <span class=file>H</span> are the
		header lines and contain information about the program that generated
		the file, the date it was created, and so on.  The line starting with
		<span class=file>S</span> begins the information about the first
		spectrum.  Following the <span class=file>S</span> is the scan number
		(twice) and the m/z of the precursor ion.  The lines beginning with
		<span class=file>Z</span> list the possible charge states of the
		spectrum (in this case 2) and the mass of the peptide at that
		charge state.  Following the <span class=file>Z</span> lines is the
		list of peaks for the spectrum.  The subsequent spectra in the file
		repeat this pattern of <span class=file>S</span> line, <span
		class=file>Z</span> line(s), and peak list.  <span
		class=file>demo.ms2</span> contains 150 spectra.
		</p>

		<a name="fasta"></a>
		<h3>Input file: protein database (fasta file)</h3>
		<p>The second input file you will need is a protein database. 
		This file is in fasta format and contains
		a list of proteins you expect to find in your sample and their
		sequences.  A sample fasta file is also comes with the distribution in
		<span class=file>doc/user/data/</span>. We will use <a
		href="../example-files/small-yeast.fasta">small-yeast.fasta</a>.  Copy it to your
		current working directory. 
		</p>

		<p>The beginning of the file looks like this.
		</p>

		<pre>
>YBL030C PET9 SGDID:S000000126, Chr II from 164000-163044, reverse complement, Verified ORF, "Major ADP/ATP carrier of the mitochondrial inner membrane, exchanges cytosolic ADP for mitochondrially synthesized ATP; required for viability in many common lab strains carrying a mutation in the polymorphic SAL1 gene"
MSSNAQVKTPLPPAPAPKKESNFLIDFLMGGVSAAVAKTAASPIERVKLLIQNQDEML
KQGTLDRKYAGILDCFKRTATQEGVISFWRGNTANVIRYFPTQALNFAFKDKIKAMFG
FKKEEGYAKWFAGNLASGGAAGALSLLFVYSLDYARTRLAADSKSSKKGGARQFNGLI
DVYKKTLKSDGVAGLYRGFLPSVVGIVVYRGLYFGMYDSLKPLLLTGSLEGSFLASFL
LGWVVTTGASTCSYPLDTVRRRMMMTSGQAVKYDGAFDCLRKIVAAEGVGSLFKGCGA
NILRGVAGAGVISMYDQLQMILFGKKFK 
		</pre>

		<p>Lines beginning with <span class=file>&gt;</span> give the name of
		a protein.  The first word is the protein name followed by an optional
		description of any length.  The following lines contain the protein
		sequence.  Proteins may or may not be separated by blank lines.
		<span class=file>small-yeast.fasta</span> contains 56 proteins.
		</p>
	<!-- END CONTENT -->
		</div>
	</div>

<footer class="site-footer">
	<div id="centerfooter">
		<div class="footerimportantlinks">
			<img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
			<ul>
				<li><a href="../faq.html">Crux <strong>FAQ</strong></a></li>
				<li><a href="../glossary.html">Glossary of terminology</a></li>
				<li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
				<li><a href="https://sourceforge.net/projects/cruxtoolkit/">SourceForge Issue's list</a></li>
				<li><a href="../release-notes.html">Release Notes</a></li>
				<li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
				<li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
				<li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
			</ul>
		</div>
		<div class="footerimportantlinks tutoriallinks">
			<img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
			<ul>
				<li><a href="install.html">Installation</a></li>
				<li><a href="gettingstarted.html">Getting started with Crux</a></li>
				<li><a href="search.html">Running a simple search using Tide and Percolator</a></li>
				<li><a href="customizedsearch.html">Customization and search options</a></li>
				<li><a href="spectralcounts.html">Using spectral-counts</a></li>
			</ul>
		</div>
		<div id="footertext">
			<p>
				The original version of Crux was written by Chris Park and Aaron Klammer
				under the supervision
				of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
				MacCoss</a>
				and <a href="http://noble.gs.washington.edu/~noble">Prof. William
				Stafford Noble</a> in the Department of Genome Sciences at the
				University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
				<br />The complete list of contributors
				can be found <a href="../contributors.html">here</a>.
				<br />
				<br />
				Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> awards R01 GM096306 and P41 GM103533. 
			</p>
		</div>
	</div>
</footer>
	
</body>

</html>
