<!DOCTYPE html>
<html>
<head>
<title>Using spectral-counts</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-V7XKGGFPYX"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-V7XKGGFPYX');
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
	// Main Menu
	$( document ).ready(function() {
			var pull 		= $('.btn');
				menu 		= $('nav ul');
				menuHeight	= menu.height();

			$(pull).on('click', function(e) {
				e.preventDefault();
				menu.slideToggle();
			});

			$(window).resize(function(){
	    		var w = $(window).width();
	    		if(w > 320 && menu.is(':hidden')) {
	    			menu.removeAttr('style');
	    		} 
			});
	});
</script>
</head>
<body>
	<div class="page-wrap">
	 	<nav>
			<div class="btn">
		    </div>
		    <img src="../images/crux-logo.png" id="logo"></a>
			<ul id="navitems">
			    <li><a href="../index.html">Home</a></li>
			    <li><a href="../download.html">Download</a></li>
			    <li><a href="../fileformats.html">File Formats</a></li>
		        <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
		    </ul>
		</nav>
		<div id="content">
		<!-- START CONTENT -->
		<h2>Using spectral-counts</h2>

		<p> 
		After you have run a <a href="search.html">search</a> and
		assigned confidence to your identified PSMs, you can now calculate
		relative quantification values for the proteins by running the
		command:<br>

		<div class=code>
		<code>$ crux spectral-counts --threshold 0.1 --protein-database small-yeast.fasta ./crux-output/percolator.target.psms.txt</code>
		</div>
		<br>

		<p>While the command is running, you will see output like this
		</p>

		<pre>
INFO: CPU: guanine.gs.washington.edu
INFO: Crux version: 4.0-ad795d65-2021-08-23
INFO: Tue Sep 14 21:35:02 PDT 2021
INFO: Beginning spectral-counts.
INFO: Total proteins found: 56
INFO: Assigning index 0 to ./crux-output/percolator.target.psms.txt.
INFO: Number of matches:9134
INFO: Number of matches passed the threshold 1515
INFO: Number of peptides 758
INFO: Number of proteins 54
INFO: Elapsed time: 0.263 s
INFO: Finished crux spectral-counts.
INFO: Return Code:0
		</pre>

		<p>
		The <span class=file>crux-output</span> directory will now contain three new files:</p>
		<ol>

		<li>spectral-counts.target.txt &ndash; tab-delimited file containing the <code>spectral-counts</code> results</li>
		<li>spectral-counts.params.txt &ndash; parameter file</li>
		<li>spectral-counts.log.txt &ndash; log file.</li>

		</ol>
		</p>

		<p>The spectral-counts.target.txt file will look something like:<br>
		<pre>
protein id	NSAF
YGR192C	0.14454275
YCR012W	0.07278429
YBR118W	0.070475481
YLL024C	0.045148738
YOR063W	0.045023989
YDR064W	0.039725326
YLR043C	0.027732428
YLR249W	0.027634142
		</pre>

		<p>
		In this output, the first field is the identifier and the second field
		is the spectral-counting measure (NSAF, dNSAF, SIN, or EMPAI) result
		for each protein.  To select which measure is computed, see the
		<a href="../commands/spectral-counts.html">crux spectral-counts</a> documentation.</p>

		<h3>Using peptideprophet probability from pepXML</h3>
		<p><code>crux spectral-counts</code> also supports
		using Peptide Prophet probabilities provided in a
		pepXML file.  Use the command:<br>
		<br>
		<div class=code>
		<code>$ crux spectral-counts --protein-database small-yeast.fasta --threshold-type custom --custom-threshold-name peptideprophet --threshold 0.9 --custom-threshold-min F target.pep.xml</code>
		</div>          
		This command will filter out matches whose Peptide Prophet probability is &lt; 0.9 and calculate NSAF for the proteins of the remaining PSMs.<br>

		<h3>Using mzIdentML files with spectral-counts</h3>
		<p><code>crux spectral-counts</code> also supports mzIdentML as input.  For example, </p> 
		<div class="code">
			$ crux spectral-counts --measure RAW file.mzid
		</div>
		<p>
		If the protein sequences are not provided within the
		given mzIdentML file, then you will have to provide
		a database using
		the <span class="code">protein-database</span>
		parameter for the NSAF, dNSAF, EMPAI, and SIN metrics.
		</p>
		<!-- END CONTENT -->
		</div>
	</div>

<footer class="site-footer">
	<div id="centerfooter">
		<div class="footerimportantlinks">
			<img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
			<ul>
				<li><a href="../faq.html">Crux FAQ</a></li>
				<li><a href="../glossary.html">Glossary of terminology</a></li>
				<li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
				<li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
				<li><a href="../release-notes.html">Release Notes</a></li>
				<li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
				<li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
				<li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
			</ul>
		</div>
		<div class="footerimportantlinks tutoriallinks">
			<img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
			<ul>
				<li><a href="install.html">Installation</a></li>
				<li><a href="gettingstarted.html">Getting started with Crux</a></li>
				<li><a href="search.html">Running a simple search using Tide and Percolator</a></li>
				<li><a href="customizedsearch.html">Customization and search options</a></li>
				<li><a href="spectralcounts.html">Using spectral-counts</a></li>
			</ul>
		</div>
		<div id="footertext">
			<p>
				The original version of Crux was written by Chris Park and Aaron Klammer
				under the supervision
				of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
				MacCoss</a>
				and <a href="http://noble.gs.washington.edu/~noble">Prof. William
				Stafford Noble</a> in the Department of Genome Sciences at the
				University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
				<br />The complete list of contributors
				can be found <a href="../contributors.html">here</a>.
				<br />
				<br />
				Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
			</p>
		</div>
	</div>
</footer>
	
</body>

</html>
