<!DOCTYPE html>
<html>
<head>
<title>Crux tandem mass spectrometry analysis software</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="styles.css"/>
<style>
        .vertical-text {
            transform: rotate(45deg);
            text-anchor:start;
        }
        .grid line {
          stroke: lightgrey;
          stroke-opacity: 0.7;
          shape-rendering: crispEdges;
        }

        .grid path {
          stroke-width: 0;
        }
        rect:hover {
          fill: orange;
        }
            
        #tooltip {
          position: absolute;
          width: auto;
          height: auto;
          padding: 10px;
          background-color: yellow;
          -webkit-border-radius: 10px;
          -moz-border-radius: 10px;
          border-radius: 10px;
          -webkit-box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);
          -moz-box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);
          box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);
          pointer-events: none;
        }

        #tooltip.hidden {
          display: none;
        }

        #tooltip p {
          margin: 0;
          font-family: Futura;
          font-size: 12px;
          line-height: 14px;
        }
</style>	
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-V7XKGGFPYX"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-V7XKGGFPYX');
</script>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
<script src="https://d3js.org/d3.v5.min.js"></script>
<script type="text/javascript">
	// Main Menu
	$( document ).ready(function() {
			var pull 		= $('.btn');
				menu 		= $('nav ul');
				menuHeight	= menu.height();

			$(pull).on('click', function(e) {
				e.preventDefault();
				menu.slideToggle();
			});

			$(window).resize(function(){
	    		var w = $(window).width();
	    		if(w > 320 && menu.is(':hidden')) {
	    			menu.removeAttr('style');
	    		} 
			});
	});
</script>
<!--(index.html only Start JS function for dynamically resizing the image map that links elements on schematic.svg-->
<script type="text/javascript">
window.onload = function () {
    var ImageMap = function (map, img) {
		var n,
			areas = map.getElementsByTagName('area'),
			len = areas.length,
			coords = [],
			previousWidth = 1125; // width of original schematic.svg
		for (n = 0; n < len; n++) {
			coords[n] = areas[n].coords.split(',');
		}
		this.resize = function () {
			var n, m, clen,
				x = img.offsetWidth / previousWidth;
			for (n = 0; n < len; n++) {
				clen = coords[n].length;
				for (m = 0; m < clen; m++) {
					coords[n][m] *= x;
				}
				areas[n].coords = coords[n].join(',');
			}
			previousWidth = img.width;
			return true;
		};
		window.onresize = this.resize;
	}

	imageMap = new ImageMap(document.getElementById('schematic_map'), document.getElementById('schematic_image'));
	imageMap.resize();
}
</script>
<!--(index.html only) End JS -->
	
<!-- Usage statistics plot -->	
<script  type="text/javascript">
        
		buildChart = function(dates, data){
		
            	var colors = ['blue', 'cyan'];
			var margin = ({top: 0, right: 0, bottom: 60, left: 50});
			var stackedData = d3.stack()
					.keys([0, 1])(data) // stacked yz
					.map((data, i) => data.map(([y0, y1]) => [y0, y1, i]));
					  
		var yMax = d3.max(data, y => d3.max(y));
		var y1Max = d3.max(stackedData, y => d3.max(y, d => d[1]))
			
		var width = 500,
		    height = 250;

		var x = d3.scaleBand()
				.domain(dates)
				.range([margin.left, width - margin.right])
                        	.padding(0.1);

            	var y = d3.scaleLinear()
				.domain([0, y1Max])
				.range([height - margin.bottom, margin.top]);

		var xAxis = svg => svg.append("g")
			.attr("transform", `translate(0,${height - margin.bottom})`)
			.call(d3.axisBottom(x)
                    .tickValues(x.domain().filter(function(d,i){ return (i%2)}))
                 )
                .selectAll("text")
                    .attr('class', 'vertical-text');
            
        var yAxisGenerator = d3.axisLeft(y).ticks(5);

        var yAxis = svg => svg.append("g")
                        .attr("transform", `translate(${margin.left},${margin.top})`)
                        .call(d3.axisLeft(y).ticks(5));

		d3.select(".chart-block")	
			.style("display", "block");
			
		var svg = d3.select(".chart")
				.attr("width", width)
				.attr("height", height);

            //-----------
            tooltipHandler = function(d, i) {
                var thisBar = d3.select(this);
        		var plot_box = d3.select(".chart")._groups[0][0].getBoundingClientRect()
                var xPosition = window.pageXOffset + plot_box.left + parseFloat(thisBar.attr("x")) + parseFloat(thisBar.attr('width')) / 2;
                var yPosition = window.pageYOffset + plot_box.top + parseFloat(thisBar.attr("y")) - 50;

                var ttBox = d3.select("#tooltip")
                    .style("left", xPosition + "px")
                    .style("top", yPosition + "px");
                
                ttBox.select("#dateField")
                     .text(dates[i]);
                ttBox.select("#percolatorField")
                     .text(data[i][1] * 1000);
                ttBox.select("#othersField")
                     .text(data[i][0] * 1000);

				d3.select("#tooltip").classed("hidden", false);
            }
            //-----------
            
		rect = svg.selectAll("g")
			.data(stackedData)
			.enter().append("g")
			  .attr("fill", (d, i) => colors[i])
			.selectAll("rect")
			.data(d => d)
			.enter().append("rect")
			  .attr("x", (d, i) => x(dates[i]))
			  .attr("y", height - margin.bottom)
			  .attr("width", x.bandwidth())
			  .attr("height", 0)
			  .on('mouseover', tooltipHandler)
			  .on('mouseout', function() {
    					d3.select("#tooltip").classed("hidden", true);
        		   })
			.transition()
				.duration(500)
				.delay((d, i) => i * 20)
				.attr("y", d => y(d[1]))
				.attr("height", d => y(d[0]) - y(d[1]));

              // add the Y gridlines
            svg.append("g")			
                .attr("class", "grid")
                .attr("transform", `translate(${margin.left},0)`)
                .call(yAxisGenerator
                      .tickSize(-(width - margin.right - margin.left))
                      .tickFormat("")
                     );
            
            svg.call(xAxis);
            svg.call(yAxis);
            
            svg.append("text")
                .attr("transform", "rotate(-90)")
                .attr("y", 0 + 0.25*margin.left)
                .attr("x",0 - (height-margin.bottom)/2)
                .attr("dy", "1em")
                .style("text-anchor", "middle")
                .style("font-size", "10px")
                .text("Crux runs (in 1,000)");            
            
            legend_data = [['percolator', colors[1]], ['others', colors[0]]];
            var legend = d3.select('svg')
                .append("g")
                .selectAll("g")
                .data(legend_data)
                .enter()
                .append('g')
                  .attr('class', 'legend')
                  .attr('transform', function(d, i) {
                    var height = 100;
                    var x = margin.left + 20;
                    var y = margin.top + 20 + i * 20;
                    return 'translate(' + x + ',' + y + ')';
            });
                      
             legend.append('rect')
                .attr('width', 15)
                .attr('height', 15)
                .style('fill', (d)=>d[1])
                .style('stroke', (d)=>d[1]);

             legend.append('text')
                .attr('x',  30)
                .attr('y', 12)
                .text((d)=>d[0]);
                      
	//-----------------------------------------------------
	}
	getData = function(){
	//-- File format: week in YYYYWW, total count, percolator count
    	//loading and parsing the data file
		$.get("gaData.csv", function(data) {
			var dates = [];
			var items = [];
			var lines = data.split('\n');
			if(lines.length > 0){
				lines.forEach(function(row) {
					var row_data = [];
					row.split(',').forEach( function(str){
							row_data.push(Number.parseInt(str));
							});
					 if(row_data.length > 1) {
						var dt_str = String(row_data[0]);
						dates.push(dt_str.substr(0, 4) + "-" + dt_str.substr(4, 2));
						items.push([(row_data[1] - row_data[2])/1000.,row_data[2]/1000.]);
					 }
				})
				buildChart(dates, items);
			}
		});
	}
</script>
	
</head>
<body  onload="getData()">
	<div class="page-wrap">
	 	<nav>
			<div class="btn">
		    </div>
		    <img src="images/crux-logo.png" id="logo"></a>
			<ul id="navitems">
			    <li><a href="index.html">Home</a></li>
			    <li><a href="download.html">Download</a></li>
			    <li><a href="fileformats.html">File Formats</a></li>
		        <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
		    </ul>
		</nav>
		<div id="content">
		<!-- START CONTENT -->
		<div id="schematic">
			<img id="schematic_image" src="images/schematic.svg" usemap="#commands-files" border="0" width="100%" alt="" />
      <!-- This doesn't work :(
      <map name="commands-files" id="schematic_map">
			  <area shape="rect" coords="253,135,349,169" alt="tide-index" href="commands/tide-index.html">
			  <area shape="rect" coords="260,247,342,278" alt="bullseye" href="commands/bullseye.html">
			  <area shape="rect" coords="510,100,612,132" alt="tide-search" href="commands/tide-search.html">
			  <area shape="rect" coords="528,191,595,222" alt="comet" href="commands/comet.html">
			  <area shape="rect" coords="512,281,613,335" alt="kojak" href="commands/kojak.html">
			  <area shape="rect" coords="776,100,873,132" alt="percolator" href="commands/percolator.html">
			  <area shape="rect" coords="784,281,867,334" alt="spectral-counts" href="commands/spectral-counts.html">
			</map>
      -->
			<img src="images/crux-logo.png" style="width:200px; float:left; padding:10px;">
			<p>The Crux mass spectrometry analysis toolkit is an open source project that aims to provide users with a cross-platform suite of analysis tools for interpreting protein mass spectrometry data.  The toolkit includes several search engines for both standard and cross-linked database search, as well as a variety of pre- and post-processing engines for assigning high-resolution precursor masses to spectra, assigning statistical confidence estimates to spectra, peptides and proteins, and performing label free quantification. Crux comes pre-complied for the Linux, Windows and MacOS operating systems. It is implemented as a single program that offers a wide variety of commands. If you use Crux in your research, please cite:
			<br />
			<span class="indentcitation">Christopher Y. Park, Aaron A. Klammer, Lukas K&auml;ll, Michael
			J. MacCoss and William Stafford
			Noble.  <a href="http://noble.gs.washington.edu/papers/park2008rapid.html">"Rapid
			and accurate peptide identification from tandem mass spectra."</a>
			<i>Journal of Proteome Research.</i> 7(7):3022-3027, 2008.</li></span>
			<br /><br />
		
			For a more up-to-date description of Crux, please read:<br />

			<span class="indentcitation">Sean McIlwain, Kaipo Tamura, Attila Kertesz-Farkas,
			Charles E.  Grant, Benjamin Diament, Barbara Frewen, J. Jeffry
			Howbert, Michael R. Hoopmann, Lukas K&auml;ll, Jimmy K. Eng, Michael
			J. MacCoss and William Stafford
			Noble.  <a href="http://pubs.acs.org/doi/abs/10.1021/pr500741y">"Crux:
			rapid open source protein tandem mass spectrometry
			analysis."</a> <i>Journal of Proteome Research.</i>
			13(10):4488-4491, 2014.</span>
			</p>
			<div class="chart-block" style="dislpay:none" align="center">
				<p>Crux weekly usage by MS proteomics community over the last 32 weeks  (in thousand runs).</p> 
				<div id="tooltip" class="hidden">
					<strong>Week:</strong><span id="dateField">100</span><br/>
					<strong>percolator:</strong><span id="percolatorField">100</span><br/>
					<strong>others:</strong><span id="othersField">100</span>
				 </div>    
				<svg class="chart"/>
			</div>
		</div>


		<br />
		<p/>
		Crux is provided as a single executable that supports the following
		commands:</p>
		<table id="commandstable">

			<tr>
			<td colspan="2"><b>Primary commands</b></td>
			</tr>

			<tr>
			<td><a href="commands/bullseye.html">bullseye</a></td>
			<td>Assign high resolution precursor m/z values to MS/MS data using the Hardkl&ouml;r algorithm.
			</tr>

			<tr>
			<td>
			<a href="commands/tide-index.html">tide-index</a></td>
			<td>
			Create an index of all peptides in a fasta file, for use in subsequent
			calls to <code>tide-search</code>.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/tide-search.html">tide-search</a></td>
			<td>
			Search a collection of spectra against a sequence database, provided
			either as a FASTA file or an index, returning a collection of
			peptide-spectrum matches (PSMs).  This is a fast search engine, but it
			runs most quickly if provided with a peptide index built
			with <code>tide-index</code>.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/cascade-search.html">cascade-search</a></td>
			<td>
			An iterative procedure for incorporating information about peptide
			groups into the database search and confidence estimation
			procedure.</td></tr>

			<tr>
			<td>
			<a href="commands/comet.html">comet</a></td>
			<td>
			Search a collection of spectra against a sequence database, returning
			a collection of PSMs.  This search engine runs directly on a protein
			database in FASTA format.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/percolator.html">percolator</a></td>
			<td>
			Re-rank and assign confidence estimates to a collection of PSMs using
			the Percolator algorithm.  Optionally, also produce protein rankings
			using the Fido algorithm.</td></tr>

			<tr>
			<td>
			<a href="commands/kojak.html">kojak</a></td>
			<td>
			Search a collection of spectra against a sequence database, finding
			cross-linked peptide matches.</td></tr>

			<tr>
			<td>
			<a href="commands/diameter.html">diameter</a></td>
			<td>
			Identify peptide sequences fromm data-independent acquisition
      mass spectrometry data without requiring a spectral library</td></tr>

			<tr>
			<td>
			<a href="commands/spectral-counts.html">spectral-counts</a></td>
			<td>
			Quantify peptides or proteins using one of three spectral counting
			methods.</td></tr>

			<tr>
			<td>
			<a href="commands/pipeline.html">pipeline</a></td>
			<td>
			Given one or more sets of tandem mass spectra as well as a protein
			database, this command runs a series of Crux tools and reports all of
			the results in a single output directory. 
			</td></tr>

			<tr>
			<td colspan="2"><b>Utilities</b></td>
			</tr>

			<tr>
			<td>
			<a href="commands/assign-confidence.html">assign-confidence</a></td>
			<td>
			Assign statistical confidence measures to each PSM in a given set.</td></tr>

			<tr>
			<td>
			<a href="commands/generate-peptides.html">generate-peptides</a></td>
			<td>
			Extract from a given set of protein sequences a list of target and
			decoy peptides fitting the specified criteria.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/get-ms2-spectrum.html">get-ms2-spectrum</a></td>
			<td>
			Extract one or more fragmentation spectra, specified by scan number, from
			an <code>MS2</code> file.</td>
			</tr>

			<tr>
			<td><a href="commands/hardklor.html">hardkl&ouml;r</a></td>
			<td>Identify isotopic distributions from high-resolution mass spectra.</td>
			</tr>

			<tr>
			<td><a href="commands/make-pin.html">make-pin</a></td>
			<td>Given a set of search results files, generate a pin file for input to
			<code>crux percolator</code></td>
			</tr>

      <tr>
      <td><a href="commands/param-medic.html">param-medic</a></td>
      <td>Examine the spectra in a file to estimate the best precursor
      and fragment error tolerances for database search.</td>
      </tr>

			<tr>
			<td>
			<a href="commands/predict-peptide-ions.html">predict-peptide-ions</a></td>
			<td>
			Given a peptide and a charge state, predict the m/z values of the
			resulting fragment ions.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/print-processed-spectra.html">print-processed-spectra</a></td>
			<td>
			Process spectra as for scoring xcorr and print the results to a file.</td></tr>

			<tr>
			<td>
			<a href="commands/psm-convert.html">psm-convert</a></td>
			<td>
			Convert a file containing peptide-spectrum matches (PSMs) from one
			format to another.</td></tr>

			<tr>
			<td>
			<a href="commands/subtract-index.html">subtract-index</a></td>
			<td>Subtract one index file from another, assuming both were generated
			by tide-index.</td></tr>

			<tr>
			<td>
			<a href="commands/localize-modification.html">localize-modification</a></td>
			<td>Finds, for each peptide-spectrum match (PSM) in a given set, the most
      most likely location along the peptide for a post-translational
      modification (PTM). The mass of the PTM is inferred from the difference
      between the spectrum neutral mass and the peptide mass.</td>
			</tr>

			<tr>
			<td>
			version</td>
			<td>
			Print the Crux version number to standard output, then exit.</td>
			</tr>

		</table>
		<!-- END CONTENT -->
		</div>
	</div>

<footer class="site-footer">
	<div id="centerfooter">
		<div class="footerimportantlinks">
			<img src="images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
			<ul>
				<li><a href="faq.html">Crux FAQ</a></li>
				<li><a href="glossary.html">Glossary of terminology</a></li>
				<li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
				<li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
				<li><a href="release-notes.html">Release Notes</a></li>
				<li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
				<li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
				<li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
			</ul>
		</div>
		<div class="footerimportantlinks tutoriallinks">
			<img src="images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
			<ul>
				<li><a href="tutorials/install.html">Installation</a></li>
				<li><a href="tutorials/gettingstarted.html">Getting started with Crux</a></li>
				<li><a href="tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
				<li><a href="tutorials/customizedsearch.html">Customization and search options</a></li>
				<li><a href="tutorials/spectralcounts.html">Using spectral-counts</a></li>
			</ul>
		</div>
		<div id="footertext">
			<p>
				The original version of Crux was written by Chris Park and Aaron Klammer
				under the supervision
				of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
				MacCoss</a>
				and <a href="http://noble.gs.washington.edu/~noble">Prof. William
				Stafford Noble</a> in the Department of Genome Sciences at the
				University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
				<br />The complete list of contributors
				can be found <a href="contributors.html">here</a>.
				<br />
				<br />
				Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
			</p>
		</div>
	</div>
</footer>
	
</body>

</html>
