<!DOCTYPE html>
<html>
<head>
<title>Crux tandem mass spectrometry analysis software</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="styles.css">
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
	// Main Menu
	$( document ).ready(function() {
			var pull 		= $('.btn');
				menu 		= $('nav ul');
				menuHeight	= menu.height();

			$(pull).on('click', function(e) {
				e.preventDefault();
				menu.slideToggle();
			});

			$(window).resize(function(){
	    		var w = $(window).width();
	    		if(w > 320 && menu.is(':hidden')) {
	    			menu.removeAttr('style');
	    		} 
			});
	});
</script>
<!--(index.html only) Start JS function for dynamically resizing the image map that links elements on schematic.png-->
<script type="text/javascript">
window.onload = function () {
    var ImageMap = function (map, img) {
		var n,
			areas = map.getElementsByTagName('area'),
			len = areas.length,
			coords = [],
			previousWidth = 1125; // width of original schematic.png
		for (n = 0; n < len; n++) {
			coords[n] = areas[n].coords.split(',');
		}
		this.resize = function () {
			var n, m, clen,
				x = img.offsetWidth / previousWidth;
			for (n = 0; n < len; n++) {
				clen = coords[n].length;
				for (m = 0; m < clen; m++) {
					coords[n][m] *= x;
				}
				areas[n].coords = coords[n].join(',');
			}
			previousWidth = img.width;
			return true;
		};
		window.onresize = this.resize;
	}

	imageMap = new ImageMap(document.getElementById('schematic_map'), document.getElementById('schematic_image'));
	imageMap.resize();
}
</script>
<!--(index.html only) End JS -->
</head>
<body>
	<div class="page-wrap">
	 	<nav>
			<div class="btn">
		    </div>
		    <img src="images/crux-logo.png" id="logo"></a>
			<ul id="navitems">
			    <li><a href="index.html">Home</a></li>
			    <li><a href="download.html">Download</a></li>
			    <li><a href="fileformats.html">File Formats</a></li>
		        <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
		    </ul>
		</nav>
		<div id="content">
		<!-- START CONTENT -->
		<div id="schematic">
			<img id="schematic_image" src="images/schematic.png" usemap="#commands-files" border="0" width="100%" alt="" />
			<map name="commands-files" id="schematic_map">
			  <area shape="rect" coords="253,135,349,169" alt="tide-index" href="commands/tide-index.html">
			  <area shape="rect" coords="260,247,342,278" alt="bullseye" href="commands/bullseye.html">
			  <area shape="rect" coords="510,100,612,132" alt="tide-search" href="commands/tide-search.html">
			  <area shape="rect" coords="528,191,595,222" alt="comet" href="commands/comet.html">
			  <area shape="rect" coords="512,281,613,335" alt="search-for-xlinks" href="commands/search-for-xlinks.html">
			  <area shape="rect" coords="776,100,873,132" alt="percolator" href="commands/percolator.html">
			  <area shape="rect" coords="791,160,861,192" alt="barista" href="commands/barista.html">
			  <area shape="rect" coords="783,222,866,252" alt="q-ranker" href="commands/q-ranker.html">
			  <area shape="rect" coords="784,281,867,334" alt="spectral-counts" href="commands/spectral-counts.html">
			</map>
			<img src="images/crux-logo.png" style="width:200px; float:left; padding:10px;">
			<p>The Crux mass spectrometry analysis toolkit is an open source project that aims to provide users with a cross-platform suite of analysis tools for interpreting protein mass spectrometry data.  The toolkit includes several search engines for both standard and cross-linked database search, as well as a variety of pre- and post-processing engines for assigning high-resolution precursor masses to spectra, assigning statistical confidence estimates to spectra, peptides and proteins, and performing label free quantification. Crux comes pre-complied for the Linux, Windows and MacOS operating systems. It is implemented as a single program that offers a wide variety of commands. If you use Crux in your research, please cite:
			<br />
			<span class="indentcitation">Christopher Y. Park, Aaron A. Klammer, Lukas K&auml;ll, Michael
			J. MacCoss and William Stafford
			Noble.  <a href="http://noble.gs.washington.edu/papers/park2008rapid.html">"Rapid
			and accurate peptide identification from tandem mass spectra."</a>
			<i>Journal of Proteome Research.</i> 7(7):3022-3027, 2008.</li></span>
			<br /><br />
		
			For a more up-to-date description of Crux, please read:<br />

			<span class="indentcitation">Sean McIlwain, Kaipo Tamura, Attila Kertesz-Farkas,
			Charles E.  Grant, Benjamin Diament, Barbara Frewen, J. Jeffry
			Howbert, Michael R. Hoopmann, Lukas K&auml;ll, Jimmy K. Eng, Michael
			J. MacCoss and William Stafford
			Noble.  <a href="http://pubs.acs.org/doi/abs/10.1021/pr500741y">"Crux:
			rapid open source protein tandem mass spectrometry
			analysis."</a> <i>Journal of Proteome Research.</i>
			13(10):4488-4491, 2014.</span>
			</p>
		</div>


		<br />
		<p/>
		Crux is provided as a single executable that supports the following
		commands:</p>
		<table id="commandstable">

			<tr>
			<td colspan="2"><b>Primary commands</b></td>
			</tr>

			<tr>
			<td><a href="commands/bullseye.html">bullseye</a></td>
			<td>Assign high resolution precursor m/z values to MS/MS data using the Hardkl&ouml;r algorithm.
			</tr>

			<tr>
			<td>
			<a href="commands/tide-index.html">tide-index</a></td>
			<td>
			Create an index of all peptides in a fasta file, for use in subsequent
			calls to <code>tide-search</code>.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/tide-search.html">tide-search</a></td>
			<td>
			Search a collection of spectra against a sequence database, provided
			either as a FASTA file or an index, returning a collection of
			peptide-spectrum matches (PSMs).  This is a fast search engine, but it
			runs most quickly if provided with a peptide index built
			with <code>tide-index</code>.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/cascade-search.html">cascade-search</a></td>
			<td>
			An iterative procedure for incorporating information about peptide
			groups into the database search and confidence estimation
			procedure.</td></tr>

			<tr>
			<td>
			<a href="commands/comet.html">comet</a></td>
			<td>
			Search a collection of spectra against a sequence database, returning
			a collection of PSMs.  This search engine runs directly on a protein
			database in FASTA format.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/percolator.html">percolator</a></td>
			<td>
			Re-rank and assign confidence estimates to a collection of PSMs using
			the Percolator algorithm.  Optionally, also produce protein rankings
			using the Fido algorithm.</td></tr>

			<tr>
			<td>
			<a href="commands/q-ranker.html">q-ranker</a></td>
			<td>
			Re-rank a collection of PSMs using the Q-ranker algorithm.</td></tr>

			<tr>
			<td>
			<a href="commands/barista.html">barista</a></td>
			<td>
			Rank PSMs, peptides and proteins, assigning a confidence measure to each
			identification.</td></tr> 

			<tr>
			<td>
			<a href="commands/search-for-xlinks.html">search-for-xlinks</a></td>
			<td>
			Search a collection of spectra against a sequence database, finding
			cross-linked peptide matches.</td></tr>

			<tr>
			<td>
			<a href="commands/spectral-counts.html">spectral-counts</a></td>
			<td>
			Quantify peptides or proteins using one of three spectral counting
			methods.</td></tr>

			<tr>
			<td>
			<a href="commands/pipeline.html">pipeline</a></td>
			<td>
			Given one or more sets of tandem mass spectra as well as a protein
			database, this command runs a series of Crux tools and reports all of
			the results in a single output directory. 
			</td></tr>

			<tr>
			<td colspan="2"><b>Utilities</b></td>
			</tr>

			<tr>
			<td>
			<a href="commands/assign-confidence.html">assign-confidence</a></td>
			<td>
			Assign statistical confidence measures to each PSM in a given set.</td></tr>

			<tr>
			<td>
			<a href="commands/generate-peptides.html">generate-peptides</a></td>
			<td>
			Extract from a given set of protein sequences a list of target and
			decoy peptides fitting the specified criteria.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/get-ms2-spectrum.html">get-ms2-spectrum</a></td>
			<td>
			Extract one or more fragmentation spectra, specified by scan number, from
			an <code>MS2</code> file.</td>
			</tr>

			<tr>
			<td><a href="commands/hardklor.html">hardkl&ouml;r</a></td>
			<td>Identify isotopic distributions from high-resolution mass spectra.</td>
			</tr>

			<tr>
			<td><a href="commands/make-pin.html">make-pin</a></td>
			<td>Given a set of search results files, generate a pin file for input to
			<code>crux percolator</code></td>
			</tr>

			<tr>
			<td>
			<a href="commands/predict-peptide-ions.html">predict-peptide-ions</a></td>
			<td>
			Given a peptide and a charge state, predict the m/z values of the
			resulting fragment ions.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/print-processed-spectra.html">print-processed-spectra</a></td>
			<td>
			Process spectra as for scoring xcorr and print the results to a file.</td></tr>

			<tr>
			<td>
			<a href="commands/psm-convert.html">psm-convert</a></td>
			<td>
			Convert a file containing peptide-spectrum matches (PSMs) from one
			format to another.</td></tr>

			<tr>
			<td>
			<a href="commands/subtract-index.html">subtract-index</a></td>
			<td>Subtract one index file from another, assuming both were generated
			by tide-index.</td></tr>

			<tr>
			<td>
			<a href="commands/xlink-assign-ions.html">xlink-assign-ions</a></td>
			<td>Given a spectrum and a pair of cross-linked peptides, assign theoretical
			ion type labels to peaks in the observed spectrum.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/xlink-score-spectrum.html">xlink-score-spectrum</a></td>
			<td>Takes a defined cross-linked peptide, a spectra file, and a scan number,
			and will calculate the XCorr score a number of different ways.</td>
			</tr>

			<tr>
			<td>
			version</td>
			<td>
			Print the Crux version number to standard output, then exit.</td>
			</tr>

			<tr>
			<td colspan="2"><b>Utilities for processing tab-delimited text files</b>
			</tr>

			<tr>
			<td>
			<a href="commands/extract-columns.html">extract-columns</a></td>
			<td>Print specified columns from a tab-delimited file.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/extract-rows.html">extract-rows</a></td>
			<td>Print specified rows from a tab-delimited file.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/stat-column.html">stat-column</a></td>
			<td>Collect summary statistics from a column in a tab-delimited file.</td>
			</tr>

			<tr>
			<td>
			<a href="commands/sort-by-column.html">sort-by-column</a></td>
			<td>Sort a tab-delimited file by a column.</td>
			</tr>

		</table>
		<!-- END CONTENT -->
		</div>
	</div>

<footer class="site-footer">
	<div id="centerfooter">
		<div class="footerimportantlinks">
			<img src="images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
			<ul>
				<li><a href="faq.html">Crux FAQ</a></li>
				<li><a href="glossary.html">Glossary of terminology</a></li>
				<li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
				<li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
				<li><a href="release-notes.html">Release Notes</a></li>
				<li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
				<li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
				<li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
			</ul>
		</div>
		<div class="footerimportantlinks tutoriallinks">
			<img src="images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
			<ul>
				<li><a href="tutorials/install.html">Installation</a></li>
				<li><a href="tutorials/gettingstarted.html">Getting started with Crux</a></li>
				<li><a href="tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
				<li><a href="tutorials/customizedsearch.html">Customization and search options</a></li>
				<li><a href="tutorials/spectralcounts.html">Using spectral-counts</a></li>
			</ul>
		</div>
		<div id="footertext">
			<p>
				The original version of Crux was written by Chris Park and Aaron Klammer
				under the supervision
				of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
				MacCoss</a>
				and <a href="http://noble.gs.washington.edu/~noble">Prof. William
				Stafford Noble</a> in the Department of Genome Sciences at the
				University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
				<br />The complete list of contributors
				can be found <a href="contributors.html">here</a>.
				<br />
				<br />
				Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
			</p>
		</div>
	</div>
</footer>
	
</body>

</html>
