<!DOCTYPE html>
<html>
<head>
<title>barista</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>
<script type="text/javascript">
  MathJax.Hub.Config({jax: ['input/TeX','output/HTML-CSS'], displayAlign: 'left'});
</script>
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);
  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
  // Main Menu
  $( document ).ready(function() {
      var pull 		= $('.btn');
        menu 		= $('nav ul');
        menuHeight	= menu.height();
      $(pull).on('click', function(e) {
        e.preventDefault();
        menu.slideToggle();
      });
      $(window).resize(function(){
          var w = $(window).width();
          if(w > 320 && menu.is(':hidden')) {
            menu.removeAttr('style');
          } 
      });
  });
</script>
</head>
<body>
  <div class="page-wrap">
    <nav>
      <div class="btn">
        </div>
        <img src="../images/crux-logo.png" id="logo"></a>
      <ul id="navitems">
          <li><a href="../index.html">Home</a></li>
          <li><a href="../download.html">Download</a></li>
          <li><a href="../fileformats.html">File Formats</a></li>
            <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
        </ul>
    </nav>
    <div id="content" class="autogenerated">
    <!-- START CONTENT -->
<h1>barista</h1>
<h2>Usage:</h2>
<p><code>crux barista [options] &lt;database&gt; &lt;fragmentation spectra&gt; &lt;search results&gt;</code></p>
<h2>Description:</h2>
<p>Barista is a protein identification algorithm that combines two different tasks&ndash;peptide-spectrum match (PSM) verification and protein inference&ndash;into a single learning algorithm. The program requires three inputs: a set of MS2 spectra, a protein database, and the results of searching the spectra against the database. Barista produces as output three ranked lists of proteins, peptides and PSMs, based on how likely the proteins and peptides are to be present in the sample and how likely the PSMs are to be correct. Barista can jointly analyze the results of multiple shotgun proteomics experiments, corresponding to different experiments or replicate runs.</p><p>Barista uses a machine learning strategy that requires that the database search be carried out on target and decoy proteins. The searches may be carried out on a concatenated database or, using the <code>--separate-searches</code> option, separate target and decoy databases. The <code>crux tide-index</code> command can be used to generate a decoy database.</p><p>Barista assigns two types of statistical confidence estimates, q-values and posterior error probabilities, to identified PSMs, peptides and proteins. For more information about these values, see the documentation for <a href="assign-confidence.html">assign-confidence</a>.</p><p>More details on the Barista algorithm are provided in</p><blockquote>Marina Spivak, Jason Weston, Michael J. MacCoss and William Stafford Noble.  <a href="http://www.mcponline.org/content/11/2/M111.012161.long">"Direct maximization of protein identifications from tandem mass spectra."</a>  <i>Molecular and Cellular Proteomics</i>. 11(2):M111.012161, 2012.</blockquote>
<h2>Input:</h2>
<ul>
  <li><code>database</code> &ndash; The program requires the FASTA format protein database files against which the search was performed. The protein database input may be a concatenated database or separate target and decoy databases; the latter is supported with the --separate-searches option, described below. In either case, Barista distinguishes between target and decoy proteins based on the presence of a decoy prefix on the sequence identifiers (see the --decoy-prefix option, below). The database can be provided in three different ways: (1) as a a single FASTA file with suffix ".fa", ".fsa" or ".fasta", (2) as a text file containing a list of FASTA files, one per line, or (3) as a directory containing multiple FASTA files (identified via the filename suffixes ".fa", ".fsa" or ".fasta").</li>
  <li><code>fragmentation spectra</code> &ndash; The fragmentation spectra must be provided in <a href="../file-formats/ms2-format.html">MS2</a>, mzXML, or MGF format.</li>
  <li><code>search results</code> &ndash; Search results in the <a href="../file-formats/txt-format.html">tab-delimited text format</a> produced by Crux or in <a href="../file-formats/sqt-format.html">SQT format</a>. Like the spectra, the search results can be provided as a single file, a list of files or a directory of files. Note, however, that the input mode for spectra and for search results must be the same; i.e., if you provide a list of files for the spectra, then you must also provide a list of files containing your search results. When the MS2 files and tab-delimited text files are provided via a file listing, it is assumed that the order of the MS2 files matches the order of the tab-delimited files. Alternatively, when the MS2 files and tab-delimited files are provided via directories, the program will search for pairs of files with the same root name but different extensions (".ms2" and ".txt").</li>
</ul>
<h2>Output:</h2>
<p>The program writes files to the folder <code>crux-output</code> by default. The name of the output folder can be set by the user using the <code>--output-dir</code> option. The following files will be created:
<ul>
  <li><code>barista.xml</code> &ndash; an <a href="../file-formats/barista-xml-format.html">XML file</a> that contains four information about proteins, subset proteins, peptides, and PSMs.</li>
  <li><code>barista.target.proteins.txt</code> &ndash; a <a href="../file-formats/txt-format.html">tab-delimited file</a> containing a ranked list of groups of indistinguishable target proteins with associated Barista scores and q-values and with peptides that contributed to the identification of the protein group.</li>
  <li><code>barista.target.subset-proteins.txt</code> &ndash; a <a href="../file-formats/txt-format.html">tab-delimited file</a> containing groups of indistinguishable proteins.</li>
  <li><code>barista.target.peptides.txt</code> &ndash; a <a href="../file-formats/txt-format.html">tab-delimited file</a> containing a ranked list of target peptides with the associated Barista scores and q-values.</li>
  <li><code>barista.target.psm.txt</code> &ndash; a <a href="../file-formats/txt-format.html">tab-delimited file format</a> containing a ranked list of target peptide-spectrum matches with the associated Barista scores and q-values.</li>
  <li><code>barista.log.txt</code> &ndash; a file where the program reports its progress.</li>
  <li><code>barista.params.txt</code> &ndash; a file with the values of all the options given to the current run.</li>
</ul>
<h2>Options:</h2>
<ul style="list-style-type: none;">
<li class="nobullet">
<h3>barista options</h3>
<ul>
  <li class="nobullet"><code>--separate-searches &lt;string&gt;</code> &ndash; If the target and decoy searches were run separately, rather than using a concatenated database, then the program will assume that the database search results provided as a required argument are from the target database search. This option then allows the user to specify the location of the decoy search results. Like the required arguments, these search results can be provided as a single file, a list of files or a directory. However, the choice (file, list or directory) must be consistent for the MS2 files and the target and decoy tab-delimited files. Also, if the MS2 and tab-delimited files are provided in directories, then Q-ranker will use the MS2 filename (foo.ms2) to identify corresponding target and decoy tab-delimited files with names like foo*.target.txt and foo*.decoy.txt. This naming convention allows the target and decoy txt files to reside in the same directory. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--skip-cleanup T|F</code> &ndash; Analysis begins with a pre-processsing step that creates a set of lookup tables which are then used during training. Normally, these lookup tables are deleted at the end of the analysis, but setting this option to T prevents the deletion of these tables. Subsequently, analyses can be repeated more efficiently by specifying the --re-run option. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--re-run &lt;string&gt;</code> &ndash; Re-run a previous analysis using a previously computed set of lookup tables. For this option to work, the --skip-cleanup option must have been set to true when the program was run the first time. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--use-spec-features T|F</code> &ndash; Use an enriched feature set, including separate features for each ion type. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--optimization protein|peptide|psm</code> &ndash; Specifies whether to do optimization at the protein, peptide or psm level. Default = <code>protein</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Enzymatic digestion</h3>
<ul>
  <li class="nobullet"><code>--enzyme no-enzyme|trypsin|trypsin/p|chymotrypsin|elastase|clostripain|cyanogen-bromide|iodosobenzoate|proline-endopeptidase|staph-protease|asp-n|lys-c|lys-n|arg-c|glu-c|pepsin-a|elastase-trypsin-chymotrypsin|lysarginase|custom-enzyme</code> &ndash; Specify the enzyme used to digest the proteins in silico. Available enzymes (with the corresponding digestion rules indicated in parentheses) include no-enzyme ([X]|[X]), trypsin ([RK]|{P}), trypsin/p ([RK]|[]), chymotrypsin ([FWYL]|{P}), elastase ([ALIV]|{P}), clostripain ([R]|[]), cyanogen-bromide ([M]|[]), iodosobenzoate ([W]|[]), proline-endopeptidase ([P]|[]), staph-protease ([E]|[]), asp-n ([]|[D]), lys-c ([K]|{P}), lys-n ([]|[K]), arg-c ([R]|{P}), glu-c ([DE]|{P}), pepsin-a ([FL]|{P}), elastase-trypsin-chymotrypsin ([ALIVKRWFY]|{P}), lysarginase ([]|[KR]). Specifying --enzyme no-enzyme yields a non-enzymatic digest. <strong>Warning:</strong> the resulting index may be quite large. Default = <code>trypsin</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Input and output</h3>
<ul>
  <li class="nobullet"><code>--decoy-prefix &lt;string&gt;</code> &ndash; Specifies the prefix of the protein names that indicate a decoy. Default = <code>decoy_</code>.</li>
  <li class="nobullet"><code>--fileroot &lt;string&gt;</code> &ndash; The fileroot string will be added as a prefix to all output file names. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--output-dir &lt;string&gt;</code> &ndash; The name of the directory where output files will be created. Default = <code>crux-output</code>.</li>
  <li class="nobullet"><code>--overwrite T|F</code> &ndash; Replace existing files if true or fail when trying to overwrite a file if false. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--pepxml-output T|F</code> &ndash; Output a pepXML results file to the output directory. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--txt-output T|F</code> &ndash; Output a tab-delimited results file to the output directory. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--parameter-file &lt;string&gt;</code> &ndash; A file containing parameters.  See the <a href="../file-formats/parameter-file.html">parameter documentation</a> page for details. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--verbosity &lt;integer&gt;</code> &ndash; Specify the verbosity of the current processes. Each level prints the following messages, including all those at lower verbosity levels: 0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on the progress of execution, 40-more progress information, 50-debug info, 60-detailed debug info. Default = <code>30</code>.</li>
  <li class="nobullet"><code>--list-of-files T|F</code> &ndash; Specify that the search results are provided as lists of files, rather than as individual files. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--feature-file-out T|F</code> &ndash; Output the computed features in <a href="../file-formats/features.html">tab-delimited Percolator input (.pin) format</a>. The features will be normalized, using either unit norm or standard deviation normalization (depending upon the value of the unit-norm option). Default = <code>false</code>.</li>
  <li class="nobullet"><code>--spectrum-parser pwiz|mstoolkit</code> &ndash; Specify the parser to use for reading in MS/MS spectra. The default, ProteoWizard parser can read the MS/MS file formats listed <a href="http://proteowizard.sourceforge.net/formats.shtml">here</a>. The alternative is <a href="../mstoolkit.html">MSToolkit parser</a>. If the ProteoWizard parser fails to read your files properly, you may want to try the MSToolkit parser instead. Default = <code>pwiz</code>.</li>
</ul>
</li>

</ul>
    <!-- END CONTENT -->
    </div>
  </div>
<footer class="site-footer">
  <div id="centerfooter">
    <div class="footerimportantlinks">
      <img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
      <ul>
        <li><a href="../faq.html">Crux FAQ</a></li>
        <li><a href="../glossary.html">Glossary of terminology</a></li>
        <li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
        <li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
        <li><a href="../release-notes.html">Release Notes</a></li>
        <li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
        <li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
        <li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
      </ul>
    </div>
    <div class="footerimportantlinks tutoriallinks">
      <img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
      <ul>
        <li><a href="../tutorials/install.html">Installation</a></li>
        <li><a href="../tutorials/gettingstarted.html">Getting started with Crux</a></li>
        <li><a href="../tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
        <li><a href="../tutorials/customizedsearch.html">Customization and search options</a></li>
        <li><a href="../tutorials/spectralcounts.html">Using spectral-counts</a></li>
      </ul>
    </div>
    <div id="footertext">
      <p>
        The original version of Crux was written by Chris Park and Aaron Klammer
        under the supervision
        of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
        MacCoss</a>
        and <a href="http://noble.gs.washington.edu/~noble">Prof. William
        Stafford Noble</a> in the Department of Genome Sciences at the
        University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
        <br />The complete list of contributors
        can be found <a href="../contributors.html">here</a>.
        <br />
        <br />
        Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
      </p>
    </div>
  </div>
</footer>
</body>
</html>