<!DOCTYPE html>
<html>
<head>
<title>cascade-search</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>
<script type="text/javascript">
  MathJax.Hub.Config({jax: ['input/TeX','output/HTML-CSS'], displayAlign: 'left'});
</script>
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);
  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
  // Main Menu
  $( document ).ready(function() {
      var pull 		= $('.btn');
        menu 		= $('nav ul');
        menuHeight	= menu.height();
      $(pull).on('click', function(e) {
        e.preventDefault();
        menu.slideToggle();
      });
      $(window).resize(function(){
          var w = $(window).width();
          if(w > 320 && menu.is(':hidden')) {
            menu.removeAttr('style');
          } 
      });
  });
</script>
</head>
<body>
  <div class="page-wrap">
    <nav>
      <div class="btn">
        </div>
        <img src="../images/crux-logo.png" id="logo"></a>
      <ul id="navitems">
          <li><a href="../index.html">Home</a></li>
          <li><a href="../download.html">Download</a></li>
          <li><a href="../fileformats.html">File Formats</a></li>
            <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
        </ul>
    </nav>
    <div id="content" class="autogenerated">
    <!-- START CONTENT -->
<h1>cascade-search</h1>
<h2>Usage:</h2>
<p><code>crux cascade-search [options] &lt;tide spectra file&gt;+ &lt;database-series&gt;</code></p>
<h2>Description:</h2>
<p>Cascade-search is a general procedure for incorporating information about peptide groups into the database search and confidence estimation procedure. Peptides may be grouped according to, for example, their enzymatic properties (zero, one, or two enzymatic termini) or the presence of different types of numbers of variable modifications. The algorithm works on a series of databases, each corresponding to a different peptide group. The database is searched in series, and after each search, any spectrum that is identified with a user-specified confidence threshold is sequestered from subsequent searches. The full cascade search procedure is described in this article: </p><blockquote>Attila Kertesz-Farkas, Uri Keich and William Stafford Noble. <a href="http://pubs.acs.org/doi/abs/10.1021/pr501173s">"Tandem mass spectrum identification via cascaded search."</a> <i>Journal of Proteome Research</i>. 14(8):3027-38, 2015. </blockquote>
<h2>Input:</h2>
<ul>
  <li><code>tide spectra file+</code> &ndash; The name of one or more files from which to parse the fragmentation spectra, in any of the file formats supported by ProteoWizard. Alternatively, the argument may be one or more binary spectrum files produced by a previous run of crux tide-search using the store-spectra parameter.</li>
  <li><code>database-series</code> &ndash; A comma-separated list of databases, each generated by tide-index. Cascade-search will search the given spectra against these databases in the given order.</li>
</ul>
<h2>Output:</h2>
<p>The program writes files to the folder <code>crux-output</code> by default. The name of the output folder can be set by the user using the <code>--output-dir</code> option. The following files will be created:
<ul>
  <li><code>cascade-search.target.txt</code> &ndash; a <a href="../file-formats/txt-format.html">tab-delimited text file</a> containing the target PSMs accepted at a pre-defined fdr.</li>
  <li><code>cascade-search.log.txt</code> &ndash; a log file containing a copy of all messages that were printed to stderr.</li>
  <li><code>cascade-search.params.txt</code> &ndash; a file containing the name and value of all parameters/options for the current operation. Not all parameters in the file may have been used in the operation. The resulting file can be used with the --parameter-file option for other crux programs.</li>
</ul>
<h2>Options:</h2>
<ul style="list-style-type: none;">
<li class="nobullet">
<h3>cascade-search options</h3>
<ul>
  <li class="nobullet"><code>--q-value-threshold &lt;float&gt;</code> &ndash; The q-value threshold used by cascade search. Each spectrum identified in one search with q-value less than this threshold will be excluded from all subsequent searches. Note that the threshold is not applied to the final database in the cascade. Default = <code>0.01</code>.</li>
  <li class="nobullet"><code>--estimation-method mix-max|tdc|peptide-level</code> &ndash; Specify the method used to estimate q-values.  The mix-max procedure or target-decoy competition apply to PSMs. The peptide-level option eliminates any PSM for which there exists a better scoring PSM involving the same peptide, and then uses decoys to assign confidence estimates. Default = <code>tdc</code>.</li>
  <li class="nobullet"><code>--score &lt;string&gt;</code> &ndash; Specify the column (for tab-delimited input) or tag (for XML input) used as input to the q-value estimation procedure. If this parameter is unspecified, then the program searches for "xcorr score", "evalue" (comet), "exact p-value" score fields in this order in the input file.  Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--sidak T|F</code> &ndash; Adjust the score using the Sidak adjustment and reports them in a new column in the output file. Note that this adjustment only makes sense if the given scores are p-values, and that it requires the presence of the "distinct matches/spectrum" feature for each PSM. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--top-match-in &lt;integer&gt;</code> &ndash; Specify the maximum rank to allow when parsing results files. Matches with ranks higher than this value will be ignored (a value of zero allows matches with any rank). Default = <code>0</code>.</li>
  <li class="nobullet"><code>--combine-charge-states T|F</code> &ndash; Specify this parameter to T in order to combine charge states with peptide sequencesin peptide-centric search. Works only if estimation-method = peptide-level. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--combine-modified-peptides T|F</code> &ndash; Specify this parameter to T in order to treat peptides carrying different or no modifications as being the same. Works only if estimation = peptide-level. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--use-tailor-calibration T|F</code> &ndash; Fast, but heuristic PSM score calibration as described in <a href="">TBA</a>. Default = <code>false</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Amino acid modifications</h3>
<ul>
  <li class="nobullet"><code>--mod-precision &lt;integer&gt;</code> &ndash; Set the precision for modifications as written to .txt files. Default = <code>2</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Precursor selection</h3>
<ul>
  <li class="nobullet"><code>--auto-precursor-window false|warn|fail</code> &ndash; Automatically estimate optimal value for the precursor-window parameter from the spectra themselves. false=no estimation, warn=try to estimate but use the default value in case of failure, fail=try to estimate and quit in case of failure. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--max-precursor-charge &lt;integer&gt;</code> &ndash; The maximum charge state of a spectra to consider in search. Default = <code>5</code>.</li>
  <li class="nobullet"><code>--precursor-window &lt;float&gt;</code> &ndash; Tolerance used for matching peptides to spectra. Peptides must be within +/- 'precursor-window' of the spectrum value. The precursor window units depend upon precursor-window-type. Default = <code>50</code>.</li>
  <li class="nobullet"><code>--precursor-window-type mass|mz|ppm</code> &ndash; Specify the units for the window that is used to select peptides around the precursor mass location (mass, mz, ppm). The magnitude of the window is defined by the precursor-window option, and candidate peptides must fall within this window. For the mass window-type, the spectrum precursor m+h value is converted to mass, and the window is defined as that mass +/- precursor-window. If the m+h value is not available, then the mass is calculated from the precursor m/z and provided charge. The peptide mass is computed as the sum of the average amino acid masses plus 18 Da for the terminal OH group. The mz window-type calculates the window as spectrum precursor m/z +/- precursor-window and then converts the resulting m/z range to the peptide mass range using the precursor charge. For the parts-per-million (ppm) window-type, the spectrum mass is calculated as in the mass type. The lower bound of the mass window is then defined as the spectrum mass / (1.0 + (precursor-window / 1000000)) and the upper bound is defined as spectrum mass / (1.0 - (precursor-window / 1000000)). Default = <code>ppm</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Search parameters</h3>
<ul>
  <li class="nobullet"><code>--auto-mz-bin-width false|warn|fail</code> &ndash; Automatically estimate optimal value for the mz-bin-width parameter from the spectra themselves. false=no estimation, warn=try to estimate but use the default value in case of failure, fail=try to estimate and quit in case of failure. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--compute-sp T|F</code> &ndash; Compute the preliminary score Sp for all candidate peptides. Report this score in the output, along with the corresponding rank, the number of matched ions and the total number of ions. This option is recommended if results are to be analyzed by Percolator or Barista. If sqt-output is enabled, then compute-sp is automatically enabled and cannot be overridden. Note that the Sp computation requires re-processing each observed spectrum, so turning on this switch involves significant computational overhead. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--deisotope &lt;float&gt;</code> &ndash; Perform a simple deisotoping operation across each MS2 spectrum. For each peak in an MS2 spectrum, consider lower m/z peaks. If the current peak occurs where an expected peak would lie for any charge state less than the charge state of the precursor, within mass tolerance, and if the current peak is of lower abundance, then the peak is removed.  The value of this parameter is the mass tolerance, in units of parts-per-million.  If set to 0, no deisotoping is performed. Default = <code>0</code>.</li>
  <li class="nobullet"><code>--exact-p-value T|F</code> &ndash; Enable the calculation of exact p-values for the XCorr score as described in <a href="http://www.ncbi.nlm.nih.gov/pubmed/24895379">this article</a>. Calculation of p-values increases the running time but increases the number of identifications at a fixed confidence threshold. The p-values will be reported in a new column with header "exact p-value", and the "xcorr score" column will be replaced with a "refactored xcorr" column. Note that, currently, p-values can only be computed when the mz-bin-width parameter is set to its default value. Variable and static mods are allowed on non-terminal residues in conjunction with p-value computation, but currently only static mods are allowed on the N-terminus, and no mods on the C-terminus. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--isotope-error &lt;string&gt;</code> &ndash; List of positive, non-zero integers. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--min-peaks &lt;integer&gt;</code> &ndash; The minimum number of peaks a spectrum must have for it to be searched. Default = <code>20</code>.</li>
  <li class="nobullet"><code>--mz-bin-offset &lt;float&gt;</code> &ndash; In the discretization of the m/z axes of the observed and theoretical spectra, this parameter specifies the location of the left edge of the first bin, relative to mass = 0 (i.e., mz-bin-offset = 0.xx means the left edge of the first bin will be located at +0.xx Da). Default = <code>0.4</code>.</li>
  <li class="nobullet"><code>--mz-bin-width &lt;float&gt;</code> &ndash; Before calculation of the XCorr score, the m/z axes of the observed and theoretical spectra are discretized. This parameter specifies the size of each bin. The exact formula for computing the discretized m/z value is floor((x/mz-bin-width) + 1.0 - mz-bin-offset), where x is the observed m/z value. For low resolution ion trap ms/ms data 1.0005079 and for high resolution ms/ms 0.02 is recommended. Default = <code>0.02</code>.</li>
  <li class="nobullet"><code>--peptide-centric-search T|F</code> &ndash; Carries out a peptide-centric search. For each peptide the top-scoring spectra are reported, in contrast to the standard spectrum-centric search where the top-scoring peptides are reported. Note that in this case the "xcorr rank" column will contain the rank of the given spectrum with respect to the given candidate peptide, rather than vice versa (which is the default). Default = <code>false</code>.</li>
  <li class="nobullet"><code>--score-function xcorr|residue-evidence|both</code> &ndash; Function used for scoring PSMs. 'xcorr' is the original scoring function used by SEQUEST; 'residue-evidence' is designed to score high-resolution MS2 spectra; and 'both' calculates both scores. The latter requires that exact-p-value=T. Default = <code>xcorr</code>.</li>
  <li class="nobullet"><code>--fragment-tolerance &lt;float&gt;</code> &ndash; Mass tolerance (in Da) for scoring pairs of peaks when creating the residue evidence matrix. This parameter only makes sense when score-function is 'residue-evidence' or 'both'. Default = <code>0.02</code>.</li>
  <li class="nobullet"><code>--evidence-granularity &lt;integer&gt;</code> &ndash; This parameter controls the granularity of the entries in the dynamic programming matrix used in residue-evidence scoring.Smaller values make the program run faster but give less accurate p-values; larger values make the program run more slowly but give more accurate p-values. Default = <code>25</code>.</li>
  <li class="nobullet"><code>--remove-precursor-peak T|F</code> &ndash; If true, all peaks around the precursor m/z will be removed, within a range specified by the --remove-precursor-tolerance option. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--remove-precursor-tolerance &lt;float&gt;</code> &ndash; This parameter specifies the tolerance (in Th) around each precursor m/z that is removed when the --remove-precursor-peak option is invoked. Default = <code>1.5</code>.</li>
  <li class="nobullet"><code>--scan-number &lt;string&gt;</code> &ndash; A single scan number or a range of numbers to be searched. Range should be specified as 'first-last' which will include scans 'first' and 'last'. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--skip-preprocessing T|F</code> &ndash; Skip preprocessing steps on spectra. Default = F. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--spectrum-charge 1|2|3|all</code> &ndash; The spectrum charges to search. With 'all' every spectrum will be searched and spectra with multiple charge states will be searched once at each charge state. With 1, 2, or 3 only spectra with that charge state will be searched. Default = <code>all</code>.</li>
  <li class="nobullet"><code>--spectrum-max-mz &lt;float&gt;</code> &ndash; The highest spectrum m/z to search in the ms2 file. Default = <code>1e+09</code>.</li>
  <li class="nobullet"><code>--spectrum-min-mz &lt;float&gt;</code> &ndash; The lowest spectrum m/z to search in the ms2 file. Default = <code>0</code>.</li>
  <li class="nobullet"><code>--use-flanking-peaks T|F</code> &ndash; Include flanking peaks around singly charged b and y theoretical ions. Each flanking peak occurs in the adjacent m/z bin and has half the intensity of the primary peak. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--use-neutral-loss-peaks T|F</code> &ndash; Controls whether neutral loss ions are considered in the search. For XCorr, the loss of ammonia (NH3, 17.0086343 Da) is applied to singly charged b- and y-ions, and the loss of water (H2O; 18.0091422) is applied to b-ions. If the precursor charge is >=3, then a doubly-charged version of each ion is added. For XCorr p-value, three types of neutral losses are included. Loss of ammonia and water are applied to b- and y-ions, and a carbon monoxide loss (CO, 27.9949) is also applied to b-ions. Higher charge fragments are included for all possible charges less than the precursor charge. All neutral loss peaks have an intensity 1/10 of the primary peak. Neutral losses are not yet implemented for the res-ev score function. Default = <code>true</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>CPU threads</h3>
<ul>
  <li class="nobullet"><code>--num-threads &lt;integer&gt;</code> &ndash; 0=poll CPU to set num threads; else specify num threads directly. Default = <code>0</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>param-medic options</h3>
<ul>
  <li class="nobullet"><code>--pm-charges &lt;string&gt;</code> &ndash; Precursor charge states to consider MS/MS spectra from, in measurement error estimation, provided as comma-separated values. Default = <code>0,2,3,4</code>.</li>
  <li class="nobullet"><code>--pm-max-frag-mz &lt;float&gt;</code> &ndash; Maximum fragment m/z value to use in measurement error estimation. Default = <code>1800</code>.</li>
  <li class="nobullet"><code>--pm-max-precursor-delta-ppm &lt;float&gt;</code> &ndash; Maximum ppm distance between precursor m/z values to consider two scans potentially generated by the same peptide for measurement error estimation. Default = <code>50</code>.</li>
  <li class="nobullet"><code>--pm-max-precursor-mz &lt;float&gt;</code> &ndash; Minimum precursor m/z value to use in measurement error estimation. Default = <code>1800</code>.</li>
  <li class="nobullet"><code>--pm-max-scan-separation &lt;integer&gt;</code> &ndash; Maximum number of scans two spectra can be separated by in order to be considered potentially generated by the same peptide, for measurement error estimation. Default = <code>1000</code>.</li>
  <li class="nobullet"><code>--pm-min-common-frag-peaks &lt;integer&gt;</code> &ndash; Number of the most-intense peaks that two spectra must share in order to potentially be generated by the same peptide, for measurement error estimation. Default = <code>20</code>.</li>
  <li class="nobullet"><code>--pm-min-frag-mz &lt;float&gt;</code> &ndash; Minimum fragment m/z value to use in measurement error estimation. Default = <code>150</code>.</li>
  <li class="nobullet"><code>--pm-min-peak-pairs &lt;integer&gt;</code> &ndash; Minimum number of peak pairs (for precursor or fragment) that must be successfully paired in order to attempt to estimate measurement error distribution. Default = <code>200</code>.</li>
  <li class="nobullet"><code>--pm-min-precursor-mz &lt;float&gt;</code> &ndash; Minimum precursor m/z value to use in measurement error estimation. Default = <code>400</code>.</li>
  <li class="nobullet"><code>--pm-min-scan-frag-peaks &lt;integer&gt;</code> &ndash; Minimum fragment peaks an MS/MS scan must contain to be used in measurement error estimation. Default = <code>40</code>.</li>
  <li class="nobullet"><code>--pm-pair-top-n-frag-peaks &lt;integer&gt;</code> &ndash; Number of fragment peaks per spectrum pair to be used in fragment error estimation. Default = <code>5</code>.</li>
  <li class="nobullet"><code>--pm-top-n-frag-peaks &lt;integer&gt;</code> &ndash; Number of most-intense fragment peaks to consider for measurement error estimation, per MS/MS spectrum. Default = <code>30</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Input and output</h3>
<ul>
  <li class="nobullet"><code>--decoy-prefix &lt;string&gt;</code> &ndash; Specifies the prefix of the protein names that indicate a decoy. Default = <code>decoy_</code>.</li>
  <li class="nobullet"><code>--verbosity &lt;integer&gt;</code> &ndash; Specify the verbosity of the current processes. Each level prints the following messages, including all those at lower verbosity levels: 0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on the progress of execution, 40-more progress information, 50-debug info, 60-detailed debug info. Default = <code>30</code>.</li>
  <li class="nobullet"><code>--parameter-file &lt;string&gt;</code> &ndash; A file containing parameters.  See the <a href="../file-formats/parameter-file.html">parameter documentation</a> page for details. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--overwrite T|F</code> &ndash; Replace existing files if true or fail when trying to overwrite a file if false. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--output-dir &lt;string&gt;</code> &ndash; The name of the directory where output files will be created. Default = <code>crux-output</code>.</li>
  <li class="nobullet"><code>--list-of-files T|F</code> &ndash; Specify that the search results are provided as lists of files, rather than as individual files. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--fileroot &lt;string&gt;</code> &ndash; The fileroot string will be added as a prefix to all output file names. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--concat T|F</code> &ndash; When set to T, target and decoy search results are reported in a single file, and only the top-scoring N matches (as specified via --top-match) are reported for each spectrum, irrespective of whether the matches involve target or decoy peptides.Note that when used with search-for-xlinks, this parameter only has an effect if use-old-xlink=F. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--file-column T|F</code> &ndash; Include the file column in tab-delimited output. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--mass-precision &lt;integer&gt;</code> &ndash; Set the precision for masses and m/z written to sqt and text files. Default = <code>4</code>.</li>
  <li class="nobullet"><code>--mzid-output T|F</code> &ndash; Output an mzIdentML results file to the output directory. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--pepxml-output T|F</code> &ndash; Output a pepXML results file to the output directory. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--pin-output T|F</code> &ndash; Output a Percolator input (PIN) file to the output directory. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--precision &lt;integer&gt;</code> &ndash; Set the precision for scores written to sqt and text files. Default = <code>8</code>.</li>
  <li class="nobullet"><code>--print-search-progress &lt;integer&gt;</code> &ndash; Show search progress by printing every n spectra searched. Set to 0 to show no search progress. Default = <code>1000</code>.</li>
  <li class="nobullet"><code>--spectrum-parser pwiz|mstoolkit</code> &ndash; Specify the parser to use for reading in MS/MS spectra. The default, ProteoWizard parser can read the MS/MS file formats listed <a href="http://proteowizard.sourceforge.net/formats.shtml">here</a>. The alternative is <a href="../mstoolkit.html">MSToolkit parser</a>. If the ProteoWizard parser fails to read your files properly, you may want to try the MSToolkit parser instead. Default = <code>pwiz</code>.</li>
  <li class="nobullet"><code>--sqt-output T|F</code> &ndash; Outputs an SQT results file to the output directory. Note that if sqt-output is enabled, then compute-sp is automatically enabled and cannot be overridden. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--store-index &lt;string&gt;</code> &ndash; When providing a FASTA file as the index, the generated binary index will be stored at the given path. This option has no effect if a binary index is provided as the index. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--store-spectra &lt;string&gt;</code> &ndash; Specify the name of the file where the binarized fragmentation spectra will be stored. Subsequent runs of crux tide-search will execute more quickly if provided with the spectra in binary format. The filename is specified relative to the current working directory, not the Crux output directory (as specified by --output-dir). This option is not valid if multiple input spectrum files are given. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--txt-output T|F</code> &ndash; Output a tab-delimited results file to the output directory. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--brief-output T|F</code> &ndash; Output in tab-delimited text only the file name, scan number, charge, score and peptide. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--use-z-line T|F</code> &ndash; Specify whether, when parsing an MS2 spectrum file, Crux obtains the precursor mass information from the "S" line or the "Z" line.  Default = <code>true</code>.</li>
</ul>
</li>

</ul>
    <!-- END CONTENT -->
    </div>
  </div>
<footer class="site-footer">
  <div id="centerfooter">
    <div class="footerimportantlinks">
      <img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
      <ul>
        <li><a href="../faq.html">Crux FAQ</a></li>
        <li><a href="../glossary.html">Glossary of terminology</a></li>
        <li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
        <li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
        <li><a href="../release-notes.html">Release Notes</a></li>
        <li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
        <li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
        <li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
      </ul>
    </div>
    <div class="footerimportantlinks tutoriallinks">
      <img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
      <ul>
        <li><a href="../tutorials/install.html">Installation</a></li>
        <li><a href="../tutorials/gettingstarted.html">Getting started with Crux</a></li>
        <li><a href="../tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
        <li><a href="../tutorials/customizedsearch.html">Customization and search options</a></li>
        <li><a href="../tutorials/spectralcounts.html">Using spectral-counts</a></li>
      </ul>
    </div>
    <div id="footertext">
      <p>
        The original version of Crux was written by Chris Park and Aaron Klammer
        under the supervision
        of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
        MacCoss</a>
        and <a href="http://noble.gs.washington.edu/~noble">Prof. William
        Stafford Noble</a> in the Department of Genome Sciences at the
        University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
        <br />The complete list of contributors
        can be found <a href="../contributors.html">here</a>.
        <br />
        <br />
        Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
      </p>
    </div>
  </div>
</footer>
</body>
</html>