<!DOCTYPE html>
<html>
<head>
<title>search-for-xlinks</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>
<script type="text/javascript">
  MathJax.Hub.Config({jax: ['input/TeX','output/HTML-CSS'], displayAlign: 'left'});
</script>
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);
  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
  // Main Menu
  $( document ).ready(function() {
      var pull 		= $('.btn');
        menu 		= $('nav ul');
        menuHeight	= menu.height();
      $(pull).on('click', function(e) {
        e.preventDefault();
        menu.slideToggle();
      });
      $(window).resize(function(){
          var w = $(window).width();
          if(w > 320 && menu.is(':hidden')) {
            menu.removeAttr('style');
          } 
      });
  });
</script>
</head>
<body>
  <div class="page-wrap">
    <nav>
      <div class="btn">
        </div>
        <img src="../images/crux-logo.png" id="logo"></a>
      <ul id="navitems">
          <li><a href="../index.html">Home</a></li>
          <li><a href="../download.html">Download</a></li>
          <li><a href="../fileformats.html">File Formats</a></li>
            <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
        </ul>
    </nav>
    <div id="content" class="autogenerated">
    <!-- START CONTENT -->
<h1>search-for-xlinks</h1>
<h2>Usage:</h2>
<p><code>crux search-for-xlinks [options] &lt;ms2 file&gt;+ &lt;protein fasta file&gt; &lt;link sites&gt; &lt;link mass&gt;</code></p>
<h2>Description:</h2>
<p>This command compares a set of spectra to cross-linked peptides derived from a protein database in FASTA format. For each spectrum, the program generates a list of candidate molecules, including linear peptides, dead-end products, self-loop products and cross-linked products, with masses that lie within a specified range of the spectrum's precursor mass. These candidate molecules are ranked using XCorr, and the XCorr scores are assigned statistical confidence estimates using an empirical curve fitting procedure.</p><p>The algorithm is described in more detail in the following article:</p><blockquote>Sean McIlwain, Paul Draghicescu, Pragya Singh, David R. Goodlett and William Stafford Noble. <a href="http://pubs.acs.org/doi/abs/10.1021/pr901163d">&quot;Detecting cross-linked peptides by searching against a database of cross-linked peptide pairs.&quot;</a> <em>Journal of Proteome Research</em>. 2010.</blockquote><p>In search-for-xlinks, properties of the cross-linker are specified using the two required command line arguments, <link sites> and <link mass>. In addition, mass shifts associated with mono-links can be specified using the --mod option. Below are suggested parameter settings for some commonly used cross-linkers:</p><table border="1">  <tr>    <td><b>Linker</b></td>    <td><b>Link Mass</b></td>    <td><b>Link Sites</b></td>    <td><b>Mono Link</b></td>  </tr>  <tr>    <td>EDC</td>    <td>-18.0156</td>    <td>K:D,E,cterm</td>    <td>&nbsp;</td>  </tr>  <tr>    <td>BS2</td></td>    <td>96.0211296</td>    <td>K,nterm:K,nterm</td>    <td>--mono-link 1K+114.0316942:T:T,1K+113.0476524:T:T</td>  </tr>  <tr>    <td>BS3</td>    <td>138.0680742</td>    <td>K,nterm:K,nterm</td>    <td>--mono-link 1K+156.0786:T:T,1K+155.0946278:T:T</td>  </tr>  <tr>    <td>DSS</td>    <td>138.0680796</td>    <td>K,nterm:K,nterm</td>    <td>--mono-link 1K+156.0786:T:T,1K+155.0946278</td>  </tr>  <tr>    <td>AMAS</td>    <td>137.011</td>    <td>K,nterm:C</td>    <td>--mono-link 1KC+155.02156:T:T</td>  </tr>  <tr>    <td>GMBS</td>    <td>165.0422</td>    <td>K,nterm:C</td>    <td>--mono-link 1KC+183.05276:T:T</td>  </tr>  <tr>    <td>formaldehyde</td>    <td>9.98435</td>    <td>K,W,nterm:H,N,Y,K,W,R,nterm</td>    <td>--mono-link 1KW+12.0:T:T,1KW+30.010565:T:T</td>  </tr></table><ul><li>Note that, unlike Tide, search-for-xlinks does not have a "decoy-format" option; instead, shuffled decoy peptides are always created.  In particular, the decoy database contains three copies of each target cross-linked species: one in which both peptides are shuffled, one in which only the first peptide is shuffled, and one in which only the second peptide is shuffled. In the tab-delimited output, these different types are indicated in the "target/decoy" column as "target-target," "target-decoy," "decoy-target" or "decoy-decoy."</li><li>In addition to the primary XCorr score, search-for-xlinks reports separate scores for the two cross-linked peptides. The way this calculation is done depends on whether the "top-n" parameter is set to 0 or not.  In the top-n=0 case, the XCorr scores of the two participating peptides are computed exactly.  When top-n is non-zero, then each peptide's score is calculated by using as the mass shift on the link site the remainder of the precursor mass. The crosslink peptide score is then calculated using the sum of the two peptide scores. In this approach, the true mass shift assigned to each peptide within a crosslinked pair can be different from that peptide's calculated mass shift. This mass shift affects the way the ions masses are calculated around the crosslink sites and will affect the final XCorr score. In particular, the smaller the precursor tolerance window, the closer the "full, slower" XCorr value calculated directly from the crosslinked peptides and the "fast, inaccurate" XCorr value calculated by summing the individual peptide scores using the remainder precursor mass shift will be. This is because the pairs of peptides chosen will have a "precursor remainder" mass shift closer to the true mass shift calculated directly from the crosslinked peptide.</li></ul>
<h2>Input:</h2>
<ul>
  <li><code>ms2 file+</code> &ndash; The name of one or more files from which to parse the fragmentation spectra, in any of the file formats supported by ProteoWizard.</li>
  <li><code>protein fasta file</code> &ndash; The name of the file in FASTA format from which to retrieve proteins.</li>
  <li><code>link sites</code> &ndash; Specification of the the two sets of amino acids that the cross-linker can connect. These are specified as two comma-separated sets of amino acids, with the two sets separated by a colon. Cross-links involving the terminus of a protein can be specified by using "nterm" or "cterm". For example, "K,nterm:Q" means that the cross linker can attach K to Q or the protein N-terminus to Q. Note that the vast majority of cross-linkers will operate on the following reactive groups: amine (K,nterm), carboxyl (D,E,cterm), sulfhydrl (C), acyl (Q) or amine+ (K,S,T,Y,nterm).</li>
  <li><code>link mass</code> &ndash; The mass modification of the linker when attached to a peptide.</li>
</ul>
<h2>Output:</h2>
<p>The program writes files to the folder <code>crux-output</code> by default. The name of the output folder can be set by the user using the <code>--output-dir</code> option. The following files will be created:
<ul>
  <li><code>search-for-xlinks.target.txt</code> &ndash; a tab-delimited text file containing the peptide-spectrum matches (PSMs). See the <a href="../file-formats/txt-format.html">txt file format</a> for a list of the fields.</li>
  <li><code>search-for-xlinks.decoy.txt</code> &ndash; a tab-delimited text file containing the decoy PSMs. See the <a href="../file-formats/txt-format.html">txt file format</a> for a list of the fields.</li>
  <li><code>search-for-xlinks.qvalues.txt</code> &ndash; a tab-delimited text file containing the top ranked PSMs with calculated q-values. See the <a href="../file-formats/txt-format.html">txt file format</a> for a list of the fields.</li>
  <li><code>search-for-xlinks.params.txt</code> &ndash; a file containing the name and value of all parameters/options for the current operation. Not all parameters in the file may have been used in the operation. The resulting file can be used with the --parameter-file option for other crux programs.</li>
  <li><code>search-for-xlinks.log.txt</code> &ndash; a log file containing a copy of all messages that were printed to stderr.</li>
</ul>
<h2>Options:</h2>
<ul style="list-style-type: none;">
<li class="nobullet">
<h3>search-for-xlinks options</h3>
<ul>
  <li class="nobullet"><code>--require-xlink-candidate T|F</code> &ndash; If there is no cross-link candidate found, then don't bother looking for linear, self-loop, and dead-link candidates. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--xlink-top-n &lt;integer&gt;</code> &ndash; Specify the number of open-mod peptides to consider in the second pass. A value of 0 will search all candiates. Default = <code>250</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Peptide properties</h3>
<ul>
  <li class="nobullet"><code>--min-mass &lt;float&gt;</code> &ndash; The minimum mass (in Da) of peptides to consider. Default = <code>200</code>.</li>
  <li class="nobullet"><code>--max-mass &lt;float&gt;</code> &ndash; The maximum mass (in Da) of peptides to consider. Default = <code>7200</code>.</li>
  <li class="nobullet"><code>--min-length &lt;integer&gt;</code> &ndash; The minimum length of peptides to consider. Default = <code>6</code>.</li>
  <li class="nobullet"><code>--max-length &lt;integer&gt;</code> &ndash; The maximum length of peptides to consider. Default = <code>50</code>.</li>
  <li class="nobullet"><code>--isotopic-mass average|mono</code> &ndash; Specify the type of isotopic masses to use when calculating the peptide mass. Default = <code>mono</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Amino acid modifications</h3>
<ul>
  <li class="nobullet"><code>--mods-spec &lt;string&gt;</code> &ndash; The general form of a modification specification has three components, as exemplified by <span style="color: red;">1</span><span style="color: green;">STY</span>+<span style="color: blue">79.966331</span>.<br>The three components are: [<span style="color: red;">max_per_peptide</span>]<span style="color: green;">residues</span>[+/-]<span style-"color: blue;">mass_change</span><br>In the example, <span style="color: red;">max_per_peptide</span> is <span style="color: red;">1</span>, <span style="color: green;">residues</span> are <span style="color: green;">STY</span>, and <span style="color: blue;">mass_change</span> is <span style="color: blue;">+79.966331</span>. To specify a static modification, the number preceding the amino acid must be omitted; i.e., <span style="color: green;">C</span>+<span style="color: blue;">57.02146</span> specifies a static modification of 57.02146 Da to cysteine. Note that Tide allows at most one modification per amino acid.  Also, the default modification (C+57.02146) will be added to every mods-spec string unless an explicit C+0 is included. Also note that search-for-xlinks allows two optional Boolean parameters with each modification, indicating whether the modification will (1) prevent enzymatic cleavage at its site, and (2) prevent cross-linking. These are specified like "K+156.0786:T:T".  By default, both of these Booleans are set to false. Default = <code>C+57.02146</code>.</li>
  <li class="nobullet"><code>--cmod &lt;string&gt;</code> &ndash; Specify a variable modification to apply to C-terminus of peptides. &lt;mass change&gt;:&lt;max distance from protein c-term (-1 for no max)&gt;. Note that this parameter only takes effect when specified in the parameter file. Default = <code>NO MODS</code>.</li>
  <li class="nobullet"><code>--nmod &lt;string&gt;</code> &ndash; Specify a variable modification to apply to N-terminus of peptides.  &lt;mass change&gt;:&lt;max distance from protein c-term (-1 for no max)&gt;. Note that this parameter only takes effect when specified in the parameter file. Default = <code>NO MODS</code>.</li>
  <li class="nobullet"><code>--max-mods &lt;integer&gt;</code> &ndash; The maximum number of modifications that can be applied to a single peptide. Default = <code>255</code>.</li>
  <li class="nobullet"><code>--mod-precision &lt;integer&gt;</code> &ndash; Set the precision for modifications as written to .txt files. Default = <code>2</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Decoy database generation</h3>
<ul>
  <li class="nobullet"><code>--seed &lt;string&gt;</code> &ndash; When given a unsigned integer value seeds the random number generator with that value. When given the string "time" seeds the random number generator with the system time. Default = <code>1</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Enzymatic digestion</h3>
<ul>
  <li class="nobullet"><code>--enzyme no-enzyme|trypsin|trypsin/p|chymotrypsin|elastase|clostripain|cyanogen-bromide|iodosobenzoate|proline-endopeptidase|staph-protease|asp-n|lys-c|lys-n|arg-c|glu-c|pepsin-a|elastase-trypsin-chymotrypsin|lysarginase|custom-enzyme</code> &ndash; Specify the enzyme used to digest the proteins in silico. Available enzymes (with the corresponding digestion rules indicated in parentheses) include no-enzyme ([X]|[X]), trypsin ([RK]|{P}), trypsin/p ([RK]|[]), chymotrypsin ([FWYL]|{P}), elastase ([ALIV]|{P}), clostripain ([R]|[]), cyanogen-bromide ([M]|[]), iodosobenzoate ([W]|[]), proline-endopeptidase ([P]|[]), staph-protease ([E]|[]), asp-n ([]|[D]), lys-c ([K]|{P}), lys-n ([]|[K]), arg-c ([R]|{P}), glu-c ([DE]|{P}), pepsin-a ([FL]|{P}), elastase-trypsin-chymotrypsin ([ALIVKRWFY]|{P}), lysarginase ([]|[KR]). Specifying --enzyme no-enzyme yields a non-enzymatic digest. <strong>Warning:</strong> the resulting index may be quite large. Default = <code>trypsin</code>.</li>
  <li class="nobullet"><code>--custom-enzyme &lt;string&gt;</code> &ndash; Specify rules for in silico digestion of protein sequences. Overrides the enzyme option. Two lists of residues are given enclosed in square brackets or curly braces and separated by a |. The first list contains residues required/prohibited before the cleavage site and the second list is residues after the cleavage site. If the residues are required for digestion, they are in square brackets, '[' and ']'. If the residues prevent digestion, then they are enclosed in curly braces, '{' and '}'. Use X to indicate all residues. For example, trypsin cuts after R or K but not before P which is represented as [RK]|{P}. AspN cuts after any residue but only before D which is represented as [X]|[D]. To prevent the sequences from being digested at all, use {X}|{X}. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--digestion full-digest|partial-digest|non-specific-digest</code> &ndash; Specify whether every peptide in the database must have two enzymatic termini (full-digest) or if peptides with only one enzymatic terminus are also included (partial-digest). Default = <code>full-digest</code>.</li>
  <li class="nobullet"><code>--missed-cleavages &lt;integer&gt;</code> &ndash; Maximum number of missed cleavages per peptide to allow in enzymatic digestion. Default = <code>0</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Precursor selection</h3>
<ul>
  <li class="nobullet"><code>--precursor-window &lt;float&gt;</code> &ndash; Tolerance used for matching peptides to spectra. Peptides must be within +/- 'precursor-window' of the spectrum value. The precursor window units depend upon precursor-window-type. Default = <code>50</code>.</li>
  <li class="nobullet"><code>--precursor-window-type mass|mz|ppm</code> &ndash; Specify the units for the window that is used to select peptides around the precursor mass location (mass, mz, ppm). The magnitude of the window is defined by the precursor-window option, and candidate peptides must fall within this window. For the mass window-type, the spectrum precursor m+h value is converted to mass, and the window is defined as that mass +/- precursor-window. If the m+h value is not available, then the mass is calculated from the precursor m/z and provided charge. The peptide mass is computed as the sum of the average amino acid masses plus 18 Da for the terminal OH group. The mz window-type calculates the window as spectrum precursor m/z +/- precursor-window and then converts the resulting m/z range to the peptide mass range using the precursor charge. For the parts-per-million (ppm) window-type, the spectrum mass is calculated as in the mass type. The lower bound of the mass window is then defined as the spectrum mass / (1.0 + (precursor-window / 1000000)) and the upper bound is defined as spectrum mass / (1.0 - (precursor-window / 1000000)). Default = <code>ppm</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Search parameters</h3>
<ul>
  <li class="nobullet"><code>--spectrum-min-mz &lt;float&gt;</code> &ndash; The lowest spectrum m/z to search in the ms2 file. Default = <code>0</code>.</li>
  <li class="nobullet"><code>--spectrum-max-mz &lt;float&gt;</code> &ndash; The highest spectrum m/z to search in the ms2 file. Default = <code>1e+09</code>.</li>
  <li class="nobullet"><code>--spectrum-charge 1|2|3|all</code> &ndash; The spectrum charges to search. With 'all' every spectrum will be searched and spectra with multiple charge states will be searched once at each charge state. With 1, 2, or 3 only spectra with that charge state will be searched. Default = <code>all</code>.</li>
  <li class="nobullet"><code>--compute-sp T|F</code> &ndash; Compute the preliminary score Sp for all candidate peptides. Report this score in the output, along with the corresponding rank, the number of matched ions and the total number of ions. This option is recommended if results are to be analyzed by Percolator or Barista. If sqt-output is enabled, then compute-sp is automatically enabled and cannot be overridden. Note that the Sp computation requires re-processing each observed spectrum, so turning on this switch involves significant computational overhead. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--precursor-window-weibull &lt;float&gt;</code> &ndash; Search decoy peptides within +/- precursor-window-weibull of the precursor mass. The resulting scores are used only for fitting the Weibull distribution Default = <code>20</code>.</li>
  <li class="nobullet"><code>--precursor-window-type-weibull mass|mz|ppm</code> &ndash; Window type to use in conjunction with the precursor-window-weibull parameter. Default = <code>mass</code>.</li>
  <li class="nobullet"><code>--min-weibull-points &lt;integer&gt;</code> &ndash; Keep shuffling and collecting XCorr scores until the minimum number of points for weibull fitting (using targets and decoys) is achieved. Default = <code>4000</code>.</li>
  <li class="nobullet"><code>--max-ion-charge &lt;string&gt;</code> &ndash; Predict theoretical ions up to max charge state (1, 2, ... ,6) or up to the charge state of the peptide ("peptide"). If the max-ion-charge is greater than the charge state of the peptide, then the maximum is the peptide charge.  Default = <code>peptide</code>.</li>
  <li class="nobullet"><code>--scan-number &lt;string&gt;</code> &ndash; A single scan number or a range of numbers to be searched. Range should be specified as 'first-last' which will include scans 'first' and 'last'. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--mz-bin-width &lt;float&gt;</code> &ndash; Before calculation of the XCorr score, the m/z axes of the observed and theoretical spectra are discretized. This parameter specifies the size of each bin. The exact formula for computing the discretized m/z value is floor((x/mz-bin-width) + 1.0 - mz-bin-offset), where x is the observed m/z value. For low resolution ion trap ms/ms data 1.0005079 and for high resolution ms/ms 0.02 is recommended. Default = <code>0.02</code>.</li>
  <li class="nobullet"><code>--mz-bin-offset &lt;float&gt;</code> &ndash; In the discretization of the m/z axes of the observed and theoretical spectra, this parameter specifies the location of the left edge of the first bin, relative to mass = 0 (i.e., mz-bin-offset = 0.xx means the left edge of the first bin will be located at +0.xx Da). Default = <code>0.4</code>.</li>
  <li class="nobullet"><code>--mod-mass-format mod-only|total|separate</code> &ndash; Specify how sequence modifications are reported in various output files. Each modification is reported as a number enclosed in square braces following the modified residue; however, the number may correspond to one of three different masses: (1) 'mod-only' reports the value of the mass shift induced by the modification; (2) 'total' reports the mass of the residue with the modification (residue mass plus modification mass); (3) 'separate' is the same as 'mod-only', but multiple modifications to a single amino acid are reported as a comma-separated list of values. For example, suppose amino acid D has an unmodified mass of 115 as well as two moifications of masses +14 and +2. In this case, the amino acid would be reported as D[16] with 'mod-only', D[131] with 'total', and D[14,2] with 'separate'. Default = <code>mod-only</code>.</li>
  <li class="nobullet"><code>--use-flanking-peaks T|F</code> &ndash; Include flanking peaks around singly charged b and y theoretical ions. Each flanking peak occurs in the adjacent m/z bin and has half the intensity of the primary peak. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--fragment-mass average|mono</code> &ndash; Specify which isotopes to use in calculating fragment ion mass. Default = <code>mono</code>.</li>
  <li class="nobullet"><code>--isotope-windows &lt;string&gt;</code> &ndash; Provides a list of isotopic windows to search. For example, -1,0,1 will search in three disjoint windows: (1) precursor_mass - neutron_mass +/- window, (2) precursor_mass +/- window, and (3) precursor_mass + neutron_mass +/- window. The window size is defined from the precursor-window and precursor-window-type parameters. This option is only available when use-old-xlink=F. Default = <code>0</code>.</li>
  <li class="nobullet"><code>--compute-p-values T|F</code> &ndash; Estimate the parameters of the score distribution for each spectrum by fitting to a Weibull distribution, and compute a p-value for each xlink product. This option is only available when use-old-xlink=F. Default = <code>false</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Fragment ion parameters</h3>
<ul>
  <li class="nobullet"><code>--use-a-ions T|F</code> &ndash; Consider a-ions in the search? Note that an a-ion is equivalent to a neutral loss of CO from the b-ion.  Peak height is 10 (in arbitrary units). Default = <code>false</code>.</li>
  <li class="nobullet"><code>--use-b-ions T|F</code> &ndash; Consider b-ions in the search? Peak height is 50 (in arbitrary units). Default = <code>true</code>.</li>
  <li class="nobullet"><code>--use-c-ions T|F</code> &ndash; Consider c-ions in the search? Peak height is 50 (in arbitrary units). Default = <code>false</code>.</li>
  <li class="nobullet"><code>--use-x-ions T|F</code> &ndash; Consider x-ions in the search? Peak height is 10 (in arbitrary units). Default = <code>false</code>.</li>
  <li class="nobullet"><code>--use-y-ions T|F</code> &ndash; Consider y-ions in the search? Peak height is 50 (in arbitrary units). Default = <code>true</code>.</li>
  <li class="nobullet"><code>--use-z-ions T|F</code> &ndash; Consider z-ions in the search? Peak height is 50 (in arbitrary units). Default = <code>false</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Cross-linking parameters</h3>
<ul>
  <li class="nobullet"><code>--use-old-xlink T|F</code> &ndash; Use the old version of xlink-searching algorithm. When false, a new version of the code is run. The new version supports variable modifications and can handle more complex databases. This new code is still in development and should be considered a beta release. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-include-linears T|F</code> &ndash; Include linear peptides in the search. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-include-deadends T|F</code> &ndash; Include dead-end peptides in the search. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-include-selfloops T|F</code> &ndash; Include self-loop peptides in the search. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-include-inter T|F</code> &ndash; Include inter-protein cross-link candidates within the search. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-include-intra T|F</code> &ndash; Include intra-protein cross-link candiates within the search. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-include-inter-intra T|F</code> &ndash; Include crosslink candidates that are both inter and intra. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--xlink-prevents-cleavage &lt;string&gt;</code> &ndash; List of amino acids for which the cross-linker can prevent cleavage. This option is only available when use-old-xlink=F. Default = <code>K</code>.</li>
  <li class="nobullet"><code>--max-xlink-mods &lt;integer&gt;</code> &ndash; Specify the maximum number of modifications allowed on a crosslinked peptide. This option is only available when use-old-xlink=F. Default = <code>255</code>.</li>
  <li class="nobullet"><code>--mono-link &lt;string&gt;</code> &ndash; Provides a list of amino acids and their mass modifications to consider as candidate for mono-/dead- links.  Format is the same as mods-spec. Default = <code>&lt;empty&gt;</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Input and output</h3>
<ul>
  <li class="nobullet"><code>--file-column T|F</code> &ndash; Include the file column in tab-delimited output. Default = <code>true</code>.</li>
  <li class="nobullet"><code>--concat T|F</code> &ndash; When set to T, target and decoy search results are reported in a single file, and only the top-scoring N matches (as specified via --top-match) are reported for each spectrum, irrespective of whether the matches involve target or decoy peptides.Note that when used with search-for-xlinks, this parameter only has an effect if use-old-xlink=F. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--spectrum-parser pwiz|mstoolkit</code> &ndash; Specify the parser to use for reading in MS/MS spectra. The default, ProteoWizard parser can read the MS/MS file formats listed <a href="http://proteowizard.sourceforge.net/formats.shtml">here</a>. The alternative is <a href="../mstoolkit.html">MSToolkit parser</a>. If the ProteoWizard parser fails to read your files properly, you may want to try the MSToolkit parser instead. Default = <code>pwiz</code>.</li>
  <li class="nobullet"><code>--use-z-line T|F</code> &ndash; Specify whether, when parsing an MS2 spectrum file, Crux obtains the precursor mass information from the "S" line or the "Z" line.  Default = <code>true</code>.</li>
  <li class="nobullet"><code>--top-match &lt;integer&gt;</code> &ndash; Specify the number of matches to report for each spectrum. Default = <code>5</code>.</li>
  <li class="nobullet"><code>--print-search-progress &lt;integer&gt;</code> &ndash; Show search progress by printing every n spectra searched. Set to 0 to show no search progress. Default = <code>1000</code>.</li>
  <li class="nobullet"><code>--output-dir &lt;string&gt;</code> &ndash; The name of the directory where output files will be created. Default = <code>crux-output</code>.</li>
  <li class="nobullet"><code>--overwrite T|F</code> &ndash; Replace existing files if true or fail when trying to overwrite a file if false. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--parameter-file &lt;string&gt;</code> &ndash; A file containing parameters.  See the <a href="../file-formats/parameter-file.html">parameter documentation</a> page for details. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--verbosity &lt;integer&gt;</code> &ndash; Specify the verbosity of the current processes. Each level prints the following messages, including all those at lower verbosity levels: 0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on the progress of execution, 40-more progress information, 50-debug info, 60-detailed debug info. Default = <code>30</code>.</li>
</ul>
</li>

</ul>
    <!-- END CONTENT -->
    </div>
  </div>
<footer class="site-footer">
  <div id="centerfooter">
    <div class="footerimportantlinks">
      <img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
      <ul>
        <li><a href="../faq.html">Crux FAQ</a></li>
        <li><a href="../glossary.html">Glossary of terminology</a></li>
        <li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
        <li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
        <li><a href="../release-notes.html">Release Notes</a></li>
        <li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
        <li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
        <li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
      </ul>
    </div>
    <div class="footerimportantlinks tutoriallinks">
      <img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
      <ul>
        <li><a href="../tutorials/install.html">Installation</a></li>
        <li><a href="../tutorials/gettingstarted.html">Getting started with Crux</a></li>
        <li><a href="../tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
        <li><a href="../tutorials/customizedsearch.html">Customization and search options</a></li>
        <li><a href="../tutorials/spectralcounts.html">Using spectral-counts</a></li>
      </ul>
    </div>
    <div id="footertext">
      <p>
        The original version of Crux was written by Chris Park and Aaron Klammer
        under the supervision
        of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
        MacCoss</a>
        and <a href="http://noble.gs.washington.edu/~noble">Prof. William
        Stafford Noble</a> in the Department of Genome Sciences at the
        University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
        <br />The complete list of contributors
        can be found <a href="../contributors.html">here</a>.
        <br />
        <br />
        Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
      </p>
    </div>
  </div>
</footer>
</body>
</html>