<!DOCTYPE html>
<html>
<head>
<title>tide-index</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<link rel="stylesheet" type="text/css" href="../styles.css">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>
<script type="text/javascript">
  MathJax.Hub.Config({jax: ['input/TeX','output/HTML-CSS'], displayAlign: 'left'});
</script>
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-26136956-1']);
  _gaq.push(['_trackPageview']);
  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
</script>
<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.8.3/jquery.min.js"></script>
<script type="text/javascript">
  // Main Menu
  $( document ).ready(function() {
      var pull 		= $('.btn');
        menu 		= $('nav ul');
        menuHeight	= menu.height();
      $(pull).on('click', function(e) {
        e.preventDefault();
        menu.slideToggle();
      });
      $(window).resize(function(){
          var w = $(window).width();
          if(w > 320 && menu.is(':hidden')) {
            menu.removeAttr('style');
          } 
      });
  });
</script>
</head>
<body>
  <div class="page-wrap">
    <nav>
      <div class="btn">
        </div>
        <img src="../images/crux-logo.png" id="logo"></a>
      <ul id="navitems">
          <li><a href="../index.html">Home</a></li>
          <li><a href="../download.html">Download</a></li>
          <li><a href="../fileformats.html">File Formats</a></li>
            <li><a href="http://groups.google.com/group/crux-users">Contact</a></li> <!--Link to google support board-->
        </ul>
    </nav>
    <div id="content" class="autogenerated">
    <!-- START CONTENT -->
<h1>tide-index</h1>
<h2>Usage:</h2>
<p><code>crux tide-index [options] &lt;protein fasta file&gt; &lt;index name&gt;</code></p>
<h2>Description:</h2>
<p>Tide is a tool for identifying peptides from tandem mass spectra. It is an independent reimplementation of the SEQUEST<sup>&reg;</sup> algorithm, which assigns peptides to spectra by comparing the observed spectra to a catalog of theoretical spectra derived from a database of known proteins. Tide's primary advantage is its speed. Our published paper provides more detail on how Tide works. If you use Tide in your research, please cite:</p><blockquote>Benjamin J. Diament and William Stafford Noble. &quot;<a href="http://dx.doi.org/10.1021/pr101196n">Faster SEQUEST Searching for Peptide Identification from Tandem Mass Spectra.</a>&quot; <em>Journal of Proteome Research</em>. 10(9):3871-9, 2011.</blockquote><p>The <code>tide-index</code> command performs an optional pre-processing step on the protein database, converting it to a binary format suitable for input to the <code>tide-search</code> command.</p><p>Tide considers only the standard set of 21 amino acids. Peptides containing non-amino acid alphanumeric characters (BJXZ) are skipped. Non-alphanumeric characters are ignored completely.</p>
<h2>Input:</h2>
<ul>
  <li><code>protein fasta file</code> &ndash; The name of the file in FASTA format from which to retrieve proteins.</li>
  <li><code>index name</code> &ndash; The desired name of the binary index.</li>
</ul>
<h2>Output:</h2>
<p>The program writes files to the folder <code>crux-output</code> by default. The name of the output folder can be set by the user using the <code>--output-dir</code> option. The following files will be created:
<ul>
  <li><code>index</code> &ndash; A binary index, using the name specified on the command line.</li>
  <li><code>tide-index.params.txt</code> &ndash; a file containing the name and value of all parameters/options for the current operation. Not all parameters in the file may have been used in the operation. The resulting file can be used with the --parameter-file option for other crux programs.</li>
  <li><code>tide-index.log.txt</code> &ndash; a log file containing a copy of all messages that were printed to the screen during execution.</li>
</ul>
<h2>Options:</h2>
<ul style="list-style-type: none;">
<li class="nobullet">
<h3>tide-index options</h3>
<ul>
  <li class="nobullet"><code>--auto-modifications-spectra &lt;string&gt;</code> &ndash; Specify the spectra file to be used for modification inference when the auto-modifications option is enabled. Multiple files may be separated by commas. Default = <code>&lt;empty&gt;</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Peptide properties</h3>
<ul>
  <li class="nobullet"><code>--clip-nterm-methionine T|F</code> &ndash; When set to T, for each protein that begins with methionine, tide-index will put two copies of the leading peptide into the index, with and without the N-terminal methionine. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--isotopic-mass average|mono</code> &ndash; Specify the type of isotopic masses to use when calculating the peptide mass. Default = <code>mono</code>.</li>
  <li class="nobullet"><code>--max-length &lt;integer&gt;</code> &ndash; The maximum length of peptides to consider. Default = <code>50</code>.</li>
  <li class="nobullet"><code>--max-mass &lt;float&gt;</code> &ndash; The maximum mass (in Da) of peptides to consider. Default = <code>7200</code>.</li>
  <li class="nobullet"><code>--min-length &lt;integer&gt;</code> &ndash; The minimum length of peptides to consider. Default = <code>6</code>.</li>
  <li class="nobullet"><code>--min-mass &lt;float&gt;</code> &ndash; The minimum mass (in Da) of peptides to consider. Default = <code>200</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Amino acid modifications</h3>
<ul>
  <li class="nobullet"><code>--cterm-peptide-mods-spec &lt;string&gt;</code> &ndash; Specify peptide c-terminal modifications. See nterm-peptide-mods-spec for syntax. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--cterm-protein-mods-spec &lt;string&gt;</code> &ndash; [[html:Specifies C-terminal static and variable mass modifications on proteins.Mod specification syntax is the same as for peptide mods (see nterm-peptide-mods-spec option),but these mods are applied only to peptide C-terminals that are also protein terminals.If variable modification are provided for both peptide and protein terminal, they will be applied one at a time.  Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--max-mods &lt;integer&gt;</code> &ndash; The maximum number of modifications that can be applied to a single peptide. Default = <code>255</code>.</li>
  <li class="nobullet"><code>--min-mods &lt;integer&gt;</code> &ndash; The minimum number of modifications that can be applied to a single peptide. Default = <code>0</code>.</li>
  <li class="nobullet"><code>--mod-precision &lt;integer&gt;</code> &ndash; Set the precision for modifications as written to .txt files. Default = <code>2</code>.</li>
  <li class="nobullet"><code>--mods-spec &lt;string&gt;</code> &ndash; The general form of a modification specification has three components, as exemplified by <span style="color: red;">1</span><span style="color: green;">STY</span>+<span style="color: blue">79.966331</span>.<br>The three components are: [<span style="color: red;">max_per_peptide</span>]<span style="color: green;">residues</span>[+/-]<span style-"color: blue;">mass_change</span><br>In the example, <span style="color: red;">max_per_peptide</span> is <span style="color: red;">1</span>, <span style="color: green;">residues</span> are <span style="color: green;">STY</span>, and <span style="color: blue;">mass_change</span> is <span style="color: blue;">+79.966331</span>. To specify a static modification, the number preceding the amino acid must be omitted; i.e., <span style="color: green;">C</span>+<span style="color: blue;">57.02146</span> specifies a static modification of 57.02146 Da to cysteine. Note that Tide allows at most one modification per amino acid.  Also, the default modification (C+57.02146) will be added to every mods-spec string unless an explicit C+0 is included. Also note that search-for-xlinks allows two optional Boolean parameters with each modification, indicating whether the modification will (1) prevent enzymatic cleavage at its site, and (2) prevent cross-linking. These are specified like "K+156.0786:T:T".  By default, both of these Booleans are set to false. Default = <code>C+57.02146</code>.</li>
  <li class="nobullet"><code>--nterm-peptide-mods-spec &lt;string&gt;</code> &ndash; Specify peptide n-terminal modifications. Like --mods-spec, this specification has three components, but with a slightly different syntax. The <span style="color: red;">max_per_peptide</span> can be either "1", in which case it defines a variable terminal modification, or missing, in which case the modification is static. The <span style="color: green;">residues</span> field indicates which amino acids are subject to the modification, with the residue <span style="color: green;">X</span> corresponding to any amino acid. Finally, <span style="color: blue;">added_mass</span> is defined as before. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--nterm-protein-mods-spec &lt;string&gt;</code> &ndash;  [[html:Same as cterm-protein-mods-spec, but for  the protein N-terminal. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--auto-modifications T|F</code> &ndash; Automatically infer modifications from the spectra themselves. Default = <code>false</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Decoy database generation</h3>
<ul>
  <li class="nobullet"><code>--allow-dups T|F</code> &ndash; Prevent duplicate peptides between the target and decoy databases. When set to "F", the program keeps all target and previously generated decoy peptides in memory. A shuffled decoy will be re-shuffled multiple times to avoid duplication. If a non-duplicated peptide cannot be generated, the decoy is skipped entirely. When set to "T", every decoy is added to the database without checking for duplication. This option reduces the memory requirements significantly. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--decoy-format none|shuffle|peptide-reverse|protein-reverse</code> &ndash; Include a decoy version of every peptide by shuffling or reversing the target sequence or protein. In shuffle or peptide-reverse mode, each peptide is either reversed or shuffled, leaving the N-terminal and C-terminal amino acids in place. Note that peptides appear multiple times in the target database are only shuffled once. In peptide-reverse mode, palindromic peptides are shuffled. Also, if a shuffled peptide produces an overlap with the target or decoy database, then the peptide is re-shuffled up to 5 times. Note that, despite this repeated shuffling, homopolymers will appear in both the target and decoy database. The protein-reverse mode reverses the entire protein sequence, irrespective of the composite peptides. Default = <code>shuffle</code>.</li>
  <li class="nobullet"><code>--keep-terminal-aminos N|C|NC|none</code> &ndash; When creating decoy peptides using decoy-format=shuffle or decoy-format=peptide-reverse, this option specifies whether the N-terminal and C-terminal amino acids are kept in place or allowed to be shuffled or reversed. For a target peptide "EAMPK" with decoy-format=peptide-reverse, setting keep-terminal-aminos to "NC" will yield "EPMAK"; setting it to "C" will yield "PMAEK"; setting it to "N" will yield "EKPMA"; and setting it to "none" will yield "KPMAE". Default = <code>NC</code>.</li>
  <li class="nobullet"><code>--num-decoys-per-target &lt;integer&gt;</code> &ndash; The number of decoys to generate per target. When set to a value n, then with concat=F tide-search will output one target and n decoys. The resulting files can be used to run the "average target-decoy competition" method in assign-confidence. This parameter only applies when decoy-format is shuffle. Default = <code>1</code>.</li>
  <li class="nobullet"><code>--seed &lt;string&gt;</code> &ndash; When given a unsigned integer value seeds the random number generator with that value. When given the string "time" seeds the random number generator with the system time. Default = <code>1</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Enzymatic digestion</h3>
<ul>
  <li class="nobullet"><code>--custom-enzyme &lt;string&gt;</code> &ndash; Specify rules for in silico digestion of protein sequences. Overrides the enzyme option. Two lists of residues are given enclosed in square brackets or curly braces and separated by a |. The first list contains residues required/prohibited before the cleavage site and the second list is residues after the cleavage site. If the residues are required for digestion, they are in square brackets, '[' and ']'. If the residues prevent digestion, then they are enclosed in curly braces, '{' and '}'. Use X to indicate all residues. For example, trypsin cuts after R or K but not before P which is represented as [RK]|{P}. AspN cuts after any residue but only before D which is represented as [X]|[D]. To prevent the sequences from being digested at all, use {X}|{X}. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--digestion full-digest|partial-digest|non-specific-digest</code> &ndash; Specify whether every peptide in the database must have two enzymatic termini (full-digest) or if peptides with only one enzymatic terminus are also included (partial-digest). Default = <code>full-digest</code>.</li>
  <li class="nobullet"><code>--enzyme no-enzyme|trypsin|trypsin/p|chymotrypsin|elastase|clostripain|cyanogen-bromide|iodosobenzoate|proline-endopeptidase|staph-protease|asp-n|lys-c|lys-n|arg-c|glu-c|pepsin-a|elastase-trypsin-chymotrypsin|lysarginase|custom-enzyme</code> &ndash; Specify the enzyme used to digest the proteins in silico. Available enzymes (with the corresponding digestion rules indicated in parentheses) include no-enzyme ([X]|[X]), trypsin ([RK]|{P}), trypsin/p ([RK]|[]), chymotrypsin ([FWYL]|{P}), elastase ([ALIV]|{P}), clostripain ([R]|[]), cyanogen-bromide ([M]|[]), iodosobenzoate ([W]|[]), proline-endopeptidase ([P]|[]), staph-protease ([E]|[]), asp-n ([]|[D]), lys-c ([K]|{P}), lys-n ([]|[K]), arg-c ([R]|{P}), glu-c ([DE]|{P}), pepsin-a ([FL]|{P}), elastase-trypsin-chymotrypsin ([ALIVKRWFY]|{P}), lysarginase ([]|[KR]). Specifying --enzyme no-enzyme yields a non-enzymatic digest. <strong>Warning:</strong> the resulting index may be quite large. Default = <code>trypsin</code>.</li>
  <li class="nobullet"><code>--missed-cleavages &lt;integer&gt;</code> &ndash; Maximum number of missed cleavages per peptide to allow in enzymatic digestion. Default = <code>0</code>.</li>
</ul>
</li>
<li class="nobullet">
<h3>Input and output</h3>
<ul>
  <li class="nobullet"><code>--decoy-prefix &lt;string&gt;</code> &ndash; Specifies the prefix of the protein names that indicate a decoy. Default = <code>decoy_</code>.</li>
  <li class="nobullet"><code>--mass-precision &lt;integer&gt;</code> &ndash; Set the precision for masses and m/z written to sqt and text files. Default = <code>4</code>.</li>
  <li class="nobullet"><code>--output-dir &lt;string&gt;</code> &ndash; The name of the directory where output files will be created. Default = <code>crux-output</code>.</li>
  <li class="nobullet"><code>--overwrite T|F</code> &ndash; Replace existing files if true or fail when trying to overwrite a file if false. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--parameter-file &lt;string&gt;</code> &ndash; A file containing parameters.  See the <a href="../file-formats/parameter-file.html">parameter documentation</a> page for details. Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--peptide-list T|F</code> &ndash; Create in the output directory a text file listing of all the peptides in the database, along with their neutral masses, one per line. If decoys are generated, then a second file will be created containing the decoy peptides. Decoys that also appear in the target database are marked with an asterisk in a third column. Default = <code>false</code>.</li>
  <li class="nobullet"><code>--temp-dir &lt;string&gt;</code> &ndash; The name of the directory where temporary files will be created. If this parameter is blank, then the system temporary directory will be used Default = <code>&lt;empty&gt;</code>.</li>
  <li class="nobullet"><code>--verbosity &lt;integer&gt;</code> &ndash; Specify the verbosity of the current processes. Each level prints the following messages, including all those at lower verbosity levels: 0-fatal errors, 10-non-fatal errors, 20-warnings, 30-information on the progress of execution, 40-more progress information, 50-debug info, 60-detailed debug info. Default = <code>30</code>.</li>
</ul>
</li>

</ul>
    <!-- END CONTENT -->
    </div>
  </div>
<footer class="site-footer">
  <div id="centerfooter">
    <div class="footerimportantlinks">
      <img src="../images/linkicon.png" style="width:16px; height:16px"><h3>Important links</h3>
      <ul>
        <li><a href="../faq.html">Crux FAQ</a></li>
        <li><a href="../glossary.html">Glossary of terminology</a></li>
        <li><a href="http://scholar.google.com/citations?hl=en&user=Rw9S1HIAAAAJ">Google Scholar profile</a></li>
        <li><a href="https://github.com/crux-toolkit/crux-toolkit">GitHub</a></li>
        <li><a href="../release-notes.html">Release Notes</a></li>
        <li><a href="https://mailman1.u.washington.edu/mailman/listinfo/crux-users" title="Receive announcements of new versions">Join the mailing list</a></li>
        <li><a href="http://www.apache.org/licenses/LICENSE-2.0">Apache license</a></li>
        <li><a href="http://groups.google.com/group/crux-users">Support Board</a></li>
      </ul>
    </div>
    <div class="footerimportantlinks tutoriallinks">
      <img src="../images/tutorialicon.png" style="height:16px"><h3>Tutorials</h3>
      <ul>
        <li><a href="../tutorials/install.html">Installation</a></li>
        <li><a href="../tutorials/gettingstarted.html">Getting started with Crux</a></li>
        <li><a href="../tutorials/search.html">Running a simple search using Tide and Percolator</a></li>
        <li><a href="../tutorials/customizedsearch.html">Customization and search options</a></li>
        <li><a href="../tutorials/spectralcounts.html">Using spectral-counts</a></li>
      </ul>
    </div>
    <div id="footertext">
      <p>
        The original version of Crux was written by Chris Park and Aaron Klammer
        under the supervision
        of <a href="http://www.gs.washington.edu/faculty/maccoss.htm">Prof. Michael
        MacCoss</a>
        and <a href="http://noble.gs.washington.edu/~noble">Prof. William
        Stafford Noble</a> in the Department of Genome Sciences at the
        University of Washington, Seattle.  Website by <a href="http://www.yuvalboss.com/">Yuval Boss</a>. 
        <br />The complete list of contributors
        can be found <a href="../contributors.html">here</a>.
        <br />
        <br />
        Maintenance and development of Crux is funded by the <a href="https://www.nih.gov/">National Institutes of Health</a> award P41 GM103533. 
      </p>
    </div>
  </div>
</footer>
</body>
</html>